---
title: 'Consistency and Replicability'
author: "Matteo Calgaro"
date: 'Compiled: `r format(Sys.Date(), "%d %B, %Y")`'
output:
  html_document:
    code_folding: hide
    df_print: kable
    number_sections: no
    theme: united
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
---

<style>
#TOC {
  top: 1%;
  opacity: 0.5;
}
#TOC:hover {
  opacity: 1;
}
</style>

```{r setup, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Data loading

Data from HMP16SData and curatedMetagenomicData

```{r}
library(HMP16SData)
library(curatedMetagenomicData)
library(plyr)
library(phyloseq)
library(reshape2)
library(ggplot2)
library(ggpubr)
library(vegan)
library(cowplot)
source("additional_functions.R")
source("../colors_17.R")
source("../eval_functions.R")
```

## 16S Datasets

```{r}
if(!file.exists("../data/16Sdatasets_for_replicability.RData")){
  # Low diversity
  A = "Supragingival Plaque"
  B = "Subgingival Plaque"
  Subgingival_Supragingival_16S <- generate_dataset_16S(A,B,seed = 2, prop = c(0.5,0.5))
  
  # Mid diversity
  A = "Attached Keratinized Gingiva"
  B = "Buccal Mucosa"
  Gingiva_Mucosa_16S <- generate_dataset_16S(A,B,seed = 1, prop = c(0.5,0.5))
  
  # High diversity
  A = "Stool"
  B = "Tongue Dorsum"
  Stool_TongueDorsum_16S <- generate_dataset_16S(A = "Stool",B = "Tongue Dorsum",seed = 1, prop = c(0.4,0.6))
  
  ps_list_16S <- list(Stool_TongueDorsum = Stool_TongueDorsum_16S,
                      Gingiva_Mucosa = Gingiva_Mucosa_16S,
                      Subgingival_Supragingival = Subgingival_Supragingival_16S)
  save(ps_list_16S,file = "../data/16Sdatasets_for_replicability.RData")
} else load("../data/16Sdatasets_for_replicability.RData")
```

## WMS Datasets

```{r}
if(!file.exists("../data/WMSdatasets_for_replicability.RData")){
  # Low diversity
  ZellerGData <- curatedMetagenomicData("ZellerG_2014.metaphlan_bugs_list*",
                      dryrun = FALSE,
                      counts = TRUE,
                      bugs.as.phyloseq = TRUE)
  psAB <- subset_samples(ZellerGData$ZellerG_2014.metaphlan_bugs_list.stool, study_condition != "adenoma")
  psAB <- prune_samples(sample_sums(psAB) >= 10^3, psAB)
  CRC_WMS <- filter_taxa(psAB,function(x) sum(x>0)>0,1)
  # Mid diversity
  Castro_NallarEData <- curatedMetagenomicData("Castro-NallarE_2015.metaphlan_bugs_list*",
                      dryrun = FALSE,
                      counts = TRUE,
                      bugs.as.phyloseq = TRUE)
  psAB <- Castro_NallarEData$`Castro-NallarE_2015.metaphlan_bugs_list.oralcavity`
  psAB <- prune_samples(sample_sums(psAB) >= 10^6, psAB)
  Schizophrenia_WMS <- filter_taxa(psAB,function(x) sum(x>0)>0,1)
  
  # High diversity 
  HMPData <- curatedMetagenomicData("HMP_2012.metaphlan_bugs_list*",
                      dryrun = FALSE,
                      counts = TRUE,
                      bugs.as.phyloseq = TRUE)
  # Selecting Tongue Dorsum
  psA <- subset_samples(HMPData$HMP_2012.metaphlan_bugs_list.oralcavity, body_subsite == "tongue_dorsum")
  # Selecting Stool
  psB <- subset_samples(HMPData$HMP_2012.metaphlan_bugs_list.stool, !duplicated(subjectID))
  # Keep one sample per individual
  psA <- subset_samples(psA, !duplicated(subjectID))
  psB <- subset_samples(psB, !duplicated(subjectID))
  # Pruning
  psA <- prune_samples(sample_sums(psA) > 10^6, psA)
  psB <- prune_samples(sample_sums(psB) > 10^6, psB)
  
  Stool_TongueDorsum_WMS <- generate_dataset_WMS(psA = psA,psB = psB,seed = 1,prop = c(0.5,0.5))
  ps_list_WMS <- list(Stool_TongueDorsum = Stool_TongueDorsum_WMS,
                      Schizophrenia = Schizophrenia_WMS,
                      CRC = CRC_WMS)
  save(ps_list_WMS,file = "../data/WMSdatasets_for_replicability.RData")
} else load("../data/WMSdatasets_for_replicability.RData")
```

# Alpha diversity

## 16S datasets

```{r, fig.width=10, fig.height=5}
plot_grid(plotlist = lapply(ps_list_16S, alpha_16S),nrow = 1,align = "h")
```

## WMS datasets

```{r, fig.width=10, fig.height=5}
plot_grid(plotlist = list(alpha_WMS(ps_list_WMS$Stool_TongueDorsum,variable = "body_subsite"),alpha_WMS(ps_list_WMS$Schizophrenia,variable = "study_condition"),alpha_WMS(ps_list_WMS$CRC,variable = "study_condition")),align = "h",nrow = 1)
```

# Beta diversity

For beta diversity and all the analyses below we use filtered phyloseq objects, e.g. features with more than 10 counts in more than 1 sample.

```{r}
if(!file.exists("../data/16Sdatasets_for_replicability_filtered.RData")){
  ps_list_16S <- lapply(ps_list_16S,function(ps){
    filter_taxa(ps,function(x) sum(x>10)>1,1)
  })
  save(ps_list_16S,file = "../data/16Sdatasets_for_replicability_filtered.RData")
} else load(file = "../data/16Sdatasets_for_replicability_filtered.RData")

if(!file.exists("../data/WMSdatasets_for_replicability_filtered.RData")){
  ps_list_WMS <- lapply(ps_list_WMS,function(ps){
    filter_taxa(ps,function(x) sum(x>10)>1,1)
  })
  save(ps_list_WMS,file = "../data/WMSdatasets_for_replicability_filtered.RData")
} else load("../data/WMSdatasets_for_replicability_filtered.RData")

cat("16S \n")
ps_list_16S
cat("WMS \n")
ps_list_WMS
```

## 16S datasets

```{r, fig.width=6, fig.height=14}
plot_grid(plotlist = list(plot_grid(plotlist = compute_MDS(ps_list_16S$Stool_TongueDorsum,normalization = "TSS",color = "HMP_BODY_SUBSITE"),nrow = 2, rel_heights = c(1,0.2)),
plot_grid(plotlist = compute_MDS(ps_list_16S$Gingiva_Mucosa,normalization = "TSS",color = "HMP_BODY_SUBSITE"),nrow = 2, rel_heights = c(1,0.2)),
plot_grid(plotlist = compute_MDS(ps_list_16S$Subgingival_Supragingival,normalization = "TSS",color = "HMP_BODY_SUBSITE"),nrow = 2, rel_heights = c(1,0.2))),nrow = 3)
```

## WMS datasets

```{r, fig.width=6, fig.height=14}
plot_grid(plotlist = list(plot_grid(plotlist = compute_MDS(ps_list_WMS$Stool_TongueDorsum,normalization = "TSS",color = "body_subsite",names = "body_subsite"),nrow = 2, rel_heights = c(1,0.2)),
plot_grid(plotlist = compute_MDS(ps_list_WMS$Schizophrenia,normalization = "TSS",color = "study_condition", names = "study_condition"),nrow = 2, rel_heights = c(1,0.2)),
plot_grid(plotlist = compute_MDS(ps_list_WMS$CRC,normalization = "TSS",color = "study_condition", names = "study_condition"),nrow = 2, rel_heights = c(1,0.2))),nrow = 3)
```

# Consistency and Replicability

To keep lighter github repository, data from each step of data preparation and DA analysis computation are not saved in the "../data" directory. Use the ready computed concordances files instead.

## Subsetting in 100 datasets

This is the code used to create reproducible subsets of the original datasets. It was necessary to repeat the random subsets generation process for each dataset instead of creating a proper method, because of issues on environmental variables when using "subset_samples(...)".

### WMS datasets

```{r, eval=FALSE}
if(!file.exists("../data/WMSsubsets_replicability.RData")){
  # low diversity
  ps = ps_list_WMS$CRC
  grp1_name = "CRC"
  grp2_name = "control"
  variable_name = "study_condition"
  # index for grp1 and grp2
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  # If one of index lists is bigger than the other, we select only the first n indexes for both lists
  # So index_1 and index_2 lists are of equal lengths
  min_length <- min(length(index_1),length(index_2))
  index_1 <- index_1[1:min_length]
  index_2 <- index_2[1:min_length]
  half_1_2 <- rep(min_length/2,2)
  names(half_1_2) <- c(grp1_name,grp2_name)
  ps <- subset_samples(ps, subjectID %in% ps@sam_data$subjectID[c(index_1,index_2)])
  # re-compute indexes for grp1 and grp2
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  ps_list <- list()
  set.seed(123)
  for(i in 1:100){
    grp1 <- sample(index_1,size = half_1_2[grp1_name],replace = FALSE)
    grp2 <- sample(index_2,size = half_1_2[grp2_name],replace = FALSE)
    ID <- list(ps@sam_data$subjectID[c(grp1,grp2)],
               ps@sam_data$subjectID[-c(grp1,grp2)])
    if(length(ID[[1]]) != length(ID[[2]])){
      ID[[2]] <- (ID[[2]])[-c(length(grp1),length(grp2)*2)]
    }
    ps_list[[i]] <- list()
    for(j in 1:length(ID)){
      ps_obj <- subset_samples(ps,subjectID %in% ID[[j]])
      ps_obj@sam_data$grp <- as.factor(ifelse(ps_obj@sam_data[,variable_name]==grp1_name,"grp1","grp2"))
      ps_obj <- filter_taxa(ps_obj,function(x) sum(x>1)>1,1)
      ps_list[[i]][[j]] <- ps_obj
    }
    names(ps_list[[i]]) <- paste0("Subset",1:2)
  }
  names(ps_list) <- paste0("Comparison",1:100)
  CRC_control <- ps_list
  
  # Mid diversity
  
  ps = ps_list_WMS$Schizophrenia
  grp1_name = "schizophrenia"
  grp2_name = "control"
  variable_name = "study_condition"
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  # If one of index lists is bigger than the other, we select only the first n indexes for both lists
  # So index_1 and index_2 lists are of equal lengths
  min_length <- min(length(index_1),length(index_2))
  index_1 <- index_1[1:min_length]
  index_2 <- index_2[1:min_length]
  half_1_2 <- rep(min_length/2,2)
  names(half_1_2) <- c(grp1_name,grp2_name)
  ps <- subset_samples(ps, subjectID %in% ps@sam_data$subjectID[c(index_1,index_2)])
  # re-compute indexes for grp1 and grp2
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  ps_list <- list()
  set.seed(123)
  for(i in 1:100){
    grp1 <- sample(index_1,size = half_1_2[grp1_name],replace = FALSE)
    grp2 <- sample(index_2,size = half_1_2[grp2_name],replace = FALSE)
    ID <- list(ps@sam_data$subjectID[c(grp1,grp2)],
               ps@sam_data$subjectID[-c(grp1,grp2)])
    if(length(ID[[1]]) != length(ID[[2]])){
      ID[[2]] <- (ID[[2]])[-c(length(grp1),length(grp2)*2)]
    }
    ps_list[[i]] <- list()
    for(j in 1:length(ID)){
      ps_obj <- subset_samples(ps,subjectID %in% ID[[j]])
      ps_obj@sam_data$grp <- as.factor(ifelse(ps_obj@sam_data[,variable_name]==grp1_name,"grp1","grp2"))
      ps_obj <- filter_taxa(ps_obj,function(x) sum(x>1)>1,1)
      ps_list[[i]][[j]] <- ps_obj
    }
    names(ps_list[[i]]) <- paste0("Subset",1:2)
  }
  names(ps_list) <- paste0("Comparison",1:100)
  schizophrenia_control <- ps_list
  
  # High diversity
  
  ps = ps_list_WMS$Stool_TongueDorsum
  grp1_name = "tongue_dorsum"
  grp2_name = "stool"
  variable_name = "body_subsite"
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  # If one of index lists is bigger than the other, we select only the first n indexes for both lists
  # So index_1 and index_2 lists are of equal lengths
  min_length <- min(length(index_1),length(index_2))
  index_1 <- index_1[1:min_length]
  index_2 <- index_2[1:min_length]
  half_1_2 <- rep(min_length/2,2)
  names(half_1_2) <- c(grp1_name,grp2_name)
  ps <- subset_samples(ps, subjectID %in% ps@sam_data$subjectID[c(index_1,index_2)])
  # re-compute indexes for grp1 and grp2
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  ps_list <- list()
  set.seed(123)
  for(i in 1:100){
    grp1 <- sample(index_1,size = half_1_2[grp1_name],replace = FALSE)
    grp2 <- sample(index_2,size = half_1_2[grp2_name],replace = FALSE)
    ID <- list(ps@sam_data$subjectID[c(grp1,grp2)],
               ps@sam_data$subjectID[-c(grp1,grp2)])
    if(length(ID[[1]]) != length(ID[[2]])){
      ID[[2]] <- (ID[[2]])[-c(length(grp1),length(grp2)*2)]
    }
    ps_list[[i]] <- list()
    for(j in 1:length(ID)){
      ps_obj <- subset_samples(ps,subjectID %in% ID[[j]])
      ps_obj@sam_data$grp <- as.factor(ifelse(ps_obj@sam_data[,variable_name]==grp1_name,"grp1","grp2"))
      ps_obj <- filter_taxa(ps_obj,function(x) sum(x>1)>1,1)
      ps_list[[i]][[j]] <- ps_obj
    }
    names(ps_list[[i]]) <- paste0("Subset",1:2)
  }
  names(ps_list) <- paste0("Comparison",1:100)
  tonguedorsum_stool <- ps_list
  
  subsets_consistency_replicability_WMS <- list(CRC_control = CRC_control,
                                                schizophrenia_control = schizophrenia_control,
                                                tonguedorsum_stool = tonguedorsum_stool)
  save(subsets_consistency_replicability_WMS,file = "../data/WMSsubsets_replicability.RData")
} else load(file = "../data/WMSsubsets_replicability.RData")
```

### 16S datasets

```{r, eval=FALSE}
if(!file.exists("../data/16Ssubsets_replicability.RData")){
  # low diversity
  ps = ps_list_16S$Subgingival_Supragingival
  ps@sam_data$subjectID <- ps@sam_data$RSID
  grp1_name = "Supragingival Plaque"
  grp2_name = "Subgingival Plaque"
  variable_name = "HMP_BODY_SUBSITE"
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  # If one of index lists is bigger than the other, we select only the first n indexes for both lists
  # So index_1 and index_2 lists are of equal lengths
  min_length <- min(length(index_1),length(index_2))
  index_1 <- index_1[1:min_length]
  index_2 <- index_2[1:min_length]
  half_1_2 <- rep(min_length/2,2)
  names(half_1_2) <- c(grp1_name,grp2_name)
  ps <- subset_samples(ps, subjectID %in% ps@sam_data$subjectID[c(index_1,index_2)])
  # re-compute indexes for grp1 and grp2
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  ps_list <- list()
  set.seed(123)
  for(i in 1:100){
    grp1 <- sample(index_1,size = half_1_2[grp1_name],replace = FALSE)
    grp2 <- sample(index_2,size = half_1_2[grp2_name],replace = FALSE)
    ID <- list(ps@sam_data$subjectID[c(grp1,grp2)],
               ps@sam_data$subjectID[-c(grp1,grp2)])
    if(length(ID[[1]]) != length(ID[[2]])){
      ID[[2]] <- (ID[[2]])[-c(length(grp1),length(grp2)*2)]
    }
    ps_list[[i]] <- list()
    for(j in 1:length(ID)){
      ps_obj <- subset_samples(ps,subjectID %in% ID[[j]])
      ps_obj@sam_data$grp <- as.factor(ifelse(ps_obj@sam_data[,variable_name]==grp1_name,"grp1","grp2"))
      ps_obj <- filter_taxa(ps_obj,function(x) sum(x>1)>1,1)
      ps_list[[i]][[j]] <- ps_obj
    }
    names(ps_list[[i]]) <- paste0("Subset",1:2)
  }
  names(ps_list) <- paste0("Comparison",1:100)
  subgingival_supragingival <- ps_list
  
  # Mid diversity
  
  ps = ps_list_16S$Gingiva_Mucosa
  ps@sam_data$subjectID <- ps@sam_data$RSID
  grp2_name = "Attached Keratinized Gingiva"
  grp1_name = "Buccal Mucosa"
  variable_name = "HMP_BODY_SUBSITE"
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  # If one of index lists is bigger than the other, we select only the first n indexes for both lists
  # So index_1 and index_2 lists are of equal lengths
  min_length <- min(length(index_1),length(index_2))
  index_1 <- index_1[1:min_length]
  index_2 <- index_2[1:min_length]
  half_1_2 <- rep(min_length/2,2)
  names(half_1_2) <- c(grp1_name,grp2_name)
  ps <- subset_samples(ps, subjectID %in% ps@sam_data$subjectID[c(index_1,index_2)])
  # re-compute indexes for grp1 and grp2
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  ps_list <- list()
  set.seed(123)
  for(i in 1:100){
    grp1 <- sample(index_1,size = half_1_2[grp1_name],replace = FALSE)
    grp2 <- sample(index_2,size = half_1_2[grp2_name],replace = FALSE)
    ID <- list(ps@sam_data$subjectID[c(grp1,grp2)],
               ps@sam_data$subjectID[-c(grp1,grp2)])
    if(length(ID[[1]]) != length(ID[[2]])){
      ID[[2]] <- (ID[[2]])[-c(length(grp1),length(grp2)*2)]
    }
    ps_list[[i]] <- list()
    for(j in 1:length(ID)){
      ps_obj <- subset_samples(ps,subjectID %in% ID[[j]])
      ps_obj@sam_data$grp <- as.factor(ifelse(ps_obj@sam_data[,variable_name]==grp1_name,"grp1","grp2"))
      ps_obj <- filter_taxa(ps_obj,function(x) sum(x>1)>1,1)
      ps_list[[i]][[j]] <- ps_obj
    }
    names(ps_list[[i]]) <- paste0("Subset",1:2)
  }
  names(ps_list) <- paste0("Comparison",1:100)
  gingiva_mucosa <- ps_list
  
  # High diversity
  
  ps = ps_list_16S$Stool_TongueDorsum
  ps@sam_data$subjectID <- ps@sam_data$RSID
  grp2_name = "Stool"
  grp1_name = "Tongue Dorsum"
  variable_name = "HMP_BODY_SUBSITE"
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  # If one of index lists is bigger than the other, we select only the first n indexes for both lists
  # So index_1 and index_2 lists are of equal lengths
  min_length <- min(length(index_1),length(index_2))
  index_1 <- index_1[1:min_length]
  index_2 <- index_2[1:min_length]
  half_1_2 <- rep(min_length/2,2)
  names(half_1_2) <- c(grp1_name,grp2_name)
  ps <- subset_samples(ps, subjectID %in% ps@sam_data$subjectID[c(index_1,index_2)])
  # re-compute indexes for grp1 and grp2
  index_1 <- which(ps@sam_data[,variable_name] == grp1_name)
  index_2 <- which(ps@sam_data[,variable_name] == grp2_name)
  ps_list <- list()
  set.seed(123)
  for(i in 1:100){
    grp1 <- sample(index_1,size = half_1_2[grp1_name],replace = FALSE)
    grp2 <- sample(index_2,size = half_1_2[grp2_name],replace = FALSE)
    ID <- list(ps@sam_data$subjectID[c(grp1,grp2)],
               ps@sam_data$subjectID[-c(grp1,grp2)])
    if(length(ID[[1]]) != length(ID[[2]])){
      ID[[2]] <- (ID[[2]])[-c(length(grp1),length(grp2)*2)]
    }
    ps_list[[i]] <- list()
    for(j in 1:length(ID)){
      ps_obj <- subset_samples(ps,subjectID %in% ID[[j]])
      ps_obj@sam_data$grp <- as.factor(ifelse(ps_obj@sam_data[,variable_name]==grp1_name,"grp1","grp2"))
      ps_obj <- filter_taxa(ps_obj,function(x) sum(x>1)>1,1)
      ps_list[[i]][[j]] <- ps_obj
    }
    names(ps_list[[i]]) <- paste0("Subset",1:2)
  }
  names(ps_list) <- paste0("Comparison",1:100)
  tonguedorsum_stool <- ps_list
  
  subsets_consistency_replicability_16S <- list(subgingival_supragingival = subgingival_supragingival,
                                                gingiva_mucosa = gingiva_mucosa,
                                                tonguedorsum_stool = tonguedorsum_stool)
  save(subsets_consistency_replicability_16S,file = "../data/16Ssubsets_replicability.RData")
} else load("../data/16Ssubsets_replicability.RData")
```

For some of the generated subsets in TongueDorsum vs Stool 16S comparisons, corncob DA detection methods will fail cause of the high sparsity and the low sample size. It is suggested to generate more than 100 subsets and replace the failed outputs. (I did it manually directly on the server).

## Differential abundance testing

### WMS

"consistency_replicability_WMS_comparison_script.R", "consistency_replicability_WMS_comparison.bash", "consistency_replicability_16S_comparison_script.R", "consistency_replicability_16S_comparison.bash" files were used to perform DA analysis using job-arrays. 

This takes a long time. On a server "songbird_data_preparation.R" run to prepare single dataset for songbird pipelines, "consistency_replicability_songbird.bash" run songbird and "consistency_replicability_aggregation.R" collected all the outputs and created a single .RData file for 16S and WMS data.

## Between Dataset Concordance and Within Dataset Concordance 

### WMS

```{r, eval=FALSE} 
# high diversity
# WMSsubsets_replicability_DA <- readRDS(file = "../data/WMSsubsets_replicability_DA_total.RDS")
# ps_fitted_list <- WMSsubsets_replicability_DA$tonguedorsum_stool
if(!file.exists("../data/WMSconcordance_tonguedorsum_stool.RData")){
  conc_df <- compute_concordance(ps_fitted_list = ps_fitted_list)
  save(conc_df,file = "../data/WMSconcordance_tonguedorsum_stool.RData")
} else load(file = "../data/WMSconcordance_tonguedorsum_stool.RData")

# mid diversity
# ps_fitted_list <- WMSsubsets_replicability_DA$schizophrenia_control
if(!file.exists("../data/WMSconcordance_schizophrenia_control.RData")){
  conc_df <- compute_concordance(ps_fitted_list = ps_fitted_list)
  save(conc_df,file = "../data/WMSconcordance_schizophrenia_control.RData")
} else load(file = "../data/WMSconcordance_schizophrenia_control.RData")

# low diversity
# ps_fitted_list <- WMSsubsets_replicability_DA$CRC_control
if(!file.exists("../data/WMSconcordance_CRC_control.RData")){
  conc_df <- compute_concordance(ps_fitted_list = ps_fitted_list)
  save(conc_df,file = "../data/WMSconcordance_CRC_control.RData")
} else load(file = "../data/WMSconcordance_CRC_control.RData")
```

### 16S

```{r, eval=FALSE} 
# high diversity
# subsets_replicability_DA_16S <- readRDS(file = "../data/16Ssubsets_replicability_DA_total.RDS")
# ps_fitted_list <- subsets_replicability_DA_16S$tonguedorsum_stool
if(!file.exists("../data/16Sconcordance_tonguedorsum_stool.RData")){
  conc_df <- compute_concordance(ps_fitted_list = ps_fitted_list)
  save(conc_df,file = "../data/16Sconcordance_tonguedorsum_stool.RData")
} else load(file = "../data/16Sconcordance_tonguedorsum_stool.RData")

# mid diversity
# ps_fitted_list <- subsets_replicability_DA_16S$gingiva_mucosa
if(!file.exists("../data/16Sconcordance_gingiva_mucosa.RData")){
  conc_df <- compute_concordance(ps_fitted_list = ps_fitted_list)
  save(conc_df,file = "../data/16Sconcordance_gingiva_mucosa.RData")
} else load(file = "../data/16Sconcordance_gingiva_mucosa.RData")

# low diversity
# ps_fitted_list <- subsets_replicability_DA_16S$subgingival_supragingival
if(!file.exists("../data/16Sconcordance_subgingival_supragingival.RData")){
  conc_df <- compute_concordance(ps_fitted_list = ps_fitted_list)
  save(conc_df,file = "../data/16Sconcordance_subgingival_supragingival.RData")
} else load(file = "../data/16Sconcordance_subgingival_supragingival.RData")
```

# Plots

```{r, fig.width=17, fig.height=11}
if(!file.exists("../data/concordance_all_datasets.RDS")){
  load("../data/WMSconcordance_tonguedorsum_stool.RData")
  conc_df_WMS_high <- conc_df
  load("../data/16Sconcordance_tonguedorsum_stool.RData")
  conc_df_16S_high <- conc_df
  
  load("../data/WMSconcordance_schizophrenia_control.RData")
  conc_df_WMS_mid <- conc_df
  load("../data/16Sconcordance_gingiva_mucosa.RData")
  conc_df_16S_mid <- conc_df
  
  load("../data/WMSconcordance_CRC_control.RData")
  conc_df_WMS_low <- conc_df
  load("../data/16Sconcordance_subgingival_supragingival.RData")
  conc_df_16S_low <- conc_df
  
  conc_df <- rbind(data.frame(conc_df_16S_high,tech = "16S",comp = "TongueDorsum_Stool",diversity = "high"),
                   data.frame(conc_df_WMS_high,tech = "WMS",comp = "TongueDorsum_Stool",diversity = "high"),
                   data.frame(conc_df_16S_mid,tech = "16S",comp = "Buccal Mucosa_Gingiva",diversity = "mid"),
                   data.frame(conc_df_WMS_mid,tech = "WMS",comp = "Schizophrenic_Control Saliva",diversity = "mid"),
                   data.frame(conc_df_16S_low,tech = "16S",comp = "Subgingival_Supragingival Plaque",diversity = "low"),
                   data.frame(conc_df_WMS_low,tech = "WMS",comp = "CRC_Control",diversity = "low"))
  
  # Some methods fail some tests but the number of feature should be always the same for each comparison and subset.
  nfeatures_correction <- suppressWarnings(ddply(conc_df, ~ comparison + subset, function(x) max(x$nfeatures)))
  
  for(i in 1:nrow(nfeatures_correction)){
    comparison <- nfeatures_correction[i,1]
    subset <- as.character(nfeatures_correction[i,2])
    nfeatures <- nfeatures_correction[i,3]
    cat("Running: ",comparison,subset,nfeatures,"\n")
    conc_df[conc_df$comparison == comparison & conc_df$subset == subset, "nfeatures"] <- nfeatures
  }
  
  concordance_df_summary <- suppressWarnings(ddply(conc_df, ~ tech + comp + diversity + rank + method1 + method2, function(x) colMeans(x[,c("concordance","nfeatures")])))
  
  saveRDS(list(concordance_df_summary = concordance_df_summary,
               conc_df = conc_df),file = "../data/concordance_all_datasets.RDS")
} else concordance_list <- readRDS(file = "../data/concordance_all_datasets.RDS")

concordance_df_summary <- concordance_list$concordance_df_summary
conc_df <- concordance_list$conc_df

AUC_AOC_between_methods <- suppressWarnings(ddply(concordance_df_summary, ~ tech + comp + diversity + method1 + method2, function(x) AucAocFun(cVals = x$concordance[x$rank <= 100], nfeatures = mean(x$nfeatures[x$rank <= 100]), threshold = 100)))
```

## Figure 4 of the manuscript

```{r, fig.width=18, fig.height=12}
a <- gheat(AUC_AOC_between_methods = AUC_AOC_between_methods,
           concordance_df_summary = concordance_df_summary,
           tech = "16S",
           comp = "TongueDorsum_Stool")

b <- gheat(AUC_AOC_between_methods = AUC_AOC_between_methods,
           concordance_df_summary = concordance_df_summary,
           tech = "WMS",
           comp = "TongueDorsum_Stool")

fig <- plot_grid(plot_grid(a$plot,b$plot,ncol = 2,labels = "auto"),
                 plot_grid(a$legend_dendro,a$legend_heat,ncol = 2,rel_widths = c(1,0.4)),nrow = 2,rel_heights = c(1,0.1))
fig

svg("../../fig4.svg",width = 18, height = 12)
fig
dev.off()
```

## Supplementary Figure - mid and low diversity datasets

```{r, fig.width=18, fig.height=24}
c <- gheat(AUC_AOC_between_methods = AUC_AOC_between_methods,
           concordance_df_summary = concordance_df_summary,
           tech = "16S",
           comp = "Buccal Mucosa_Gingiva")

d <- gheat(AUC_AOC_between_methods = AUC_AOC_between_methods,
           concordance_df_summary = concordance_df_summary,
           tech = "WMS",
           comp = "Schizophrenic_Control Saliva")

e <- gheat(AUC_AOC_between_methods = AUC_AOC_between_methods,
           concordance_df_summary = concordance_df_summary,
           tech = "16S",
           comp = "Subgingival_Supragingival Plaque")

f <- gheat(AUC_AOC_between_methods = AUC_AOC_between_methods,
           concordance_df_summary = concordance_df_summary,
           tech = "WMS",
           comp = "CRC_Control")

fig <- plot_grid(plot_grid(c$plot,d$plot,e$plot,f$plot,ncol = 2,labels = "auto"),
                 plot_grid(c$legend_dendro,c$legend_heat,ncol = 2,rel_widths = c(1,0.4)),nrow = 2,rel_heights = c(1,0.1))
fig

svg("../../fig4_mid_low.svg",width = 18, height = 24)
fig
dev.off()
```


```{r, fig.width=11, fig.height=12}

a5 <- g_AUC(conc_df = conc_df,div = "high",tech = "16S")
b5 <- g_AUC(conc_df = conc_df,div = "high",tech = "WMS")
c5 <- g_AUC(conc_df = conc_df,div = "mid",tech = "16S")
d5 <- g_AUC(conc_df = conc_df,div = "mid",tech = "WMS")
e5 <- g_AUC(conc_df = conc_df,div = "low",tech = "16S")
f5 <- g_AUC(conc_df = conc_df,div = "low",tech = "WMS")

abcde <- plot_grid(a5,b5,c5,d5,e5,f5,nrow = 3,ncol = 2,labels = "auto")
fig <- plot_grid(abcde,a$legend_dendro,nrow = 2,rel_heights = c(1,0.1))
fig

svg(filename = "../../fig5.svg", width = 11, height = 12)
fig
dev.off()
```

# Summary

Average WMC values between datasets and simulations to rank methods.

```{r}
concordance_list <- readRDS(file = "../data/concordance_all_datasets.RDS")
concordance_df_summary <- concordance_list$concordance_df_summary

conc_df <- concordance_list$conc_df
AUC_WM_df <- conc_df[conc_df$subset == "1vs2" & conc_df$rank == 100,]

summary_rank <- function(AUC_WM_df = AUC_WM_df,tech,diversity){
  data <- AUC_WM_df[AUC_WM_df$tech == tech & AUC_WM_df$diversity == diversity,]
  ranking <- ddply(data, ~ comparison, function(x) rank(-x[,"concordance"])*(16/nrow(x)))
  colnames(ranking)[2:ncol(ranking)] <- as.character(unique(data$method1))
  return(data.frame("method" = as.character(unique(data$method1)),
                    "value" = colMeans(ranking[2:ncol(ranking)])))
}

rank_WM_concordance <- data.frame(rbind(summary_rank(AUC_WM_df,"16S","high"),
                                        summary_rank(AUC_WM_df,"16S","mid"),
                                        summary_rank(AUC_WM_df,"16S","low"),
                                        summary_rank(AUC_WM_df,"WMS","high"),
                                        summary_rank(AUC_WM_df,"WMS","mid"),
                                        summary_rank(AUC_WM_df,"WMS","low")))

rank_WM_concordance_summary <- ddply(rank_WM_concordance, .variables = ~ method, function(x) mean(x[,"value"]))

colnames(rank_WM_concordance_summary) <- c("method","value")

saveRDS(file = "../data/summary/WMC_summary.RDS",object = rank_WM_concordance_summary)
```





