---
title: 'Enrichment analysis on real data'
author: "Matteo Calgaro"
date: 'Compiled: `r format(Sys.Date(), "%d %B, %Y")`'
output:
  html_document:
    code_folding: hide
    df_print: kable
    number_sections: no
    theme: united
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
---

<style>
#TOC {
  top: 1%;
  opacity: 0.5;
}
#TOC:hover {
  opacity: 1;
}
</style>

```{r setup, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE,message=FALSE,warning=FALSE}
library(phyloseq)
library(ggforce)
library(cowplot)
library(HotLoadings)
source("../eval_functions.R")
```

# Data loading

Biological knowledge about microbial niches has been collected in literature for some of the datasets used in the concordance analysis:
<ul>
  <li>https://www.ncbi.nlm.nih.gov/pubmed/31076212</li>
  <li>https://github.com/waldronlab/nychanesmicrobiome/blob/master/inst/extdata/biosis.tsv</li>
</ul>

```{r, echo=FALSE,warning=FALSE,message=FALSE}
library(curatedMetagenomicData)
# Extraction from HMP Project
HMPData <- curatedMetagenomicData("HMP_2012.metaphlan_bugs_list*",
                    dryrun = FALSE,
                    counts = TRUE,
                    bugs.as.phyloseq = TRUE)
# Selecting Subgingival
psA <- subset_samples(HMPData$HMP_2012.metaphlan_bugs_list.oralcavity, body_subsite == "subgingival_plaque")
# Selecting Supragingival
psB <- subset_samples(HMPData$HMP_2012.metaphlan_bugs_list.oralcavity, body_subsite == "supragingival_plaque")
# Keep one sample per individual
psA <- subset_samples(psA, !duplicated(subjectID))
psB <- subset_samples(psB, !duplicated(subjectID))
set.seed(123)
selected <- sample(1:nsamples(psB),size = nsamples(psA),replace = FALSE)
sel <- vector(mode = "logical",nsamples(psB))
sel[selected] <- TRUE
psB <- subset_samples(psB,sel)
# Pruning
psA <- prune_samples(sample_sums(psA) > 10^6, psA)
psB <- prune_samples(sample_sums(psB) > 10^6, psB)

ps <- merge_phyloseq(psA,psB)
ps <- filter_taxa(ps,function(x) sum(x>10)>1,1)
  
genera_methabolism <- read.table(file = "../data/genera_methabolism.tsv",sep = "\t",header = TRUE)
```

# Supragingival vs Subgingival Plaque

It is of interest to perform an enrichment analysis on the lists of differential abundant genera to study the aerobic or anaerobic methabolism of genera colonizing supragingival and subgingival plaque.

```{r}
grp1 = "subgingival_plaque"
grp2 = "supragingival_plaque"
#ps <- ps_list_16S$Subgingival_Supragingival

ps@sam_data$grp = factor(ps@sam_data$body_subsite,levels = c(grp1,grp2),labels = c("grp1","grp2"))

# Genera metadata about methabolism
methabolism_df <- data.frame("Genus" = ps@tax_table@.Data[,"Genus"],row.names = names(ps@tax_table@.Data[,"Genus"]))
index_meth <- match(methabolism_df$Genus,genera_methabolism$Genera)
methabolism_df$Meth <- genera_methabolism$Meth[index_meth]

if(!file.exists("../data/supragingival_subgingival_DA_WMS.RData")){
  eval_methods <- oneSimRunGSOwn(physeq = ps,epsilon = 1e14, 
                                 grid.keepX = c(seq(from = ntaxa(ps)*0.05,to = ntaxa(ps),by = 10)))
  
  if(!file.exists("../data/supragingival_subgingival_DA_WMS_corncob_bootstrap.RDS")){
    corncob_LRT_bootstrap <- corncobmodel(ps,test = "LRT",bootstrap = TRUE,B = 100)
    corncob_wald_bootstrap <- corncobmodel(ps,test = "Wald", bootstrap = TRUE, B = 100)
    saveRDS(object = list("corncob_LRT_bootstrap" = corncob_LRT_bootstrap,
                          "corncob_wald_bootstrap" = corncob_wald_bootstrap),
            file = "../data/supragingival_subgingival_DA_WMS_corncob_bootstrap.RDS")
  } else corncob_bootstrap <- readRDS("../data/supragingival_subgingival_DA_WMS_corncob_bootstrap.RDS")
  
  eval_methods$corncob_LRT_bootstrap <- corncob_bootstrap$corncob_LRT_bootstrap
  eval_methods$corncob_wald_bootstrap <- corncob_bootstrap$corncob_wald_bootstrap

  ### Songbird data preparation
  ## OTU table
  write.table(x = "#OTU ID\t",
              file = "./songbird/WMS/supragingival_subgingival_otutable.tsv",
              quote = FALSE,
              row.names = FALSE, 
              col.names = FALSE,
              eol = "")
  write.table(x = ps@otu_table, append = TRUE,
              file = "./songbird/WMS/supragingival_subgingival_otutable.tsv",
              quote = FALSE,
              sep = "\t",
              row.names = TRUE,
              col.names = TRUE)
  ## Sample data
  write.table(x = "sampleID\t",
              file = "./songbird/WMS/supragingival_subgingival_samdata.tsv",
              quote = FALSE,
              row.names = FALSE, 
              col.names = FALSE,
              eol = "")
  write.table(x = ps@sam_data, 
              append = TRUE,
              file = "./songbird/WMS/supragingival_subgingival_samdata.tsv",
              quote = FALSE,
              sep = "\t",
              row.names = TRUE,
              col.names = TRUE)
  
  ## Then switch to unix bash to run songbird:
  # module load Anaconda/2019.07
  # source activate songbird_env
  # biom convert -i './songbird/supragingival_subgingival_WMS/supragingival_subgingival_otutable.tsv' -o './songbird/supragingival_subgingival_WMS/supragingival_subgingival_otutable.biom' --to-hdf5
  # songbird multinomial --input-biom './songbird/supragingival_subgingival_WMS/supragingival_subgingival_otutable.biom' \
  #                      --metadata-file './songbird/supragingival_subgingival_WMS/supragingival_subgingival_samdata.tsv' \
  #                      --formula "grp" \
  #                      --summary-dir './songbird/supragingival_subgingival_WMS/'
  ## This procedure generates the "differentials.tsv" file
  
  songbird <- read.table(file = "./songbird/WMS/songbird/differentials.tsv",header = TRUE)
  eval_methods$songbird$statInfo <- data.frame(songbird,row.names = songbird$featureid)
  
  save(eval_methods, file = "../data/supragingival_subgingival_DA_WMS.RData")
} else load(file = "../data/supragingival_subgingival_DA_WMS.RData")


```

Many methods, many outputs:
<ul>
  <li>MAST, edgeR, limma and DESeq2 based methods have "logFC" column in statInfo matrix;</li>
  <li>metagenomeSeq has "coef_grp2" column of statInfo matrix which represents the FC;</li>
  <li>scde has "ce" column of statInfo matrix which represents the conservative estimate of expression-fold change (equals to the min(abs(c(lb, ub))), or 0 if the CI crosses the 0;</li>
  <li>seurat has "avg_logFC" column in statInfo matrix;</li>
  <li>ALDEx2 has "effect" column in statInfo matrix which represents a vector containing the per-feature effect size (grp1 vs grp2);</li>
  <li>corncob methods has "mu.grpgrp2" coefficient which represents the FC in statInfo matrix;</li>
  <li>mixMC has the "value.var" column which represents the loading;</li>
  <li>songbird has the "grp.T.grp2." column which contains the differentials for grp2.</li>
</ul>

For most methods a \code{pValMat} object is present. We use statInfo information to assess the differential abundance (DA) direction, while we use pValMat to count the number of DA features (corrected p-values).

```{r}
col_names <- c("logFC","coef_grp2","mu.grpgrp2","mle","avg_logFC","effect","value.var","grp.T.grp2.","coef")

extract_statistics <- function(method, col_names = col_names, alpha = 0.05, percent = 5, methabolism_df = methabolism_df, ps){
  which_col <- which(col_names %in% colnames(method$statInfo))
  if("pValMat" %in% names(method)){# for most methods
    
    # pValMat investigation for DA identification
    DA_features_index <- which(method$pValMat[,"adjP"] < alpha)
    DA_features <- names(DA_features_index)
    if(is.null(rownames(method$statInfo)))
      rownames(method$statInfo) <- rownames(method$pValMat)
    
  } else if(col_names[which_col] == "grp.T.grp2."){# songbird
    
    differentials <- method$statInfo[,c("featureid",col_names[which_col])]
    ordered_index <- order(-abs(differentials[,col_names[which_col]]))
    DA_features_index <- ordered_index[1:round(ntaxa(ps)*percent/100)] # Top % of total taxa
    DA_features <- differentials[DA_features_index,"featureid"]
    
  } else if(col_names[which_col] == "value.var"){# mixMC
    
    DA_features <- rownames(method$statInfo)
    # the latent variable has inverse sign in WMS
    method$statInfo$value.var <- -method$statInfo$value.var 
    
  } 
  
  if(col_names[which_col] == "coef"){# corncob methods have melted statInfo matrix
    
    which_col <- which(col_names %in% method$statInfo$coef)
    method$statInfo <- method$statInfo[method$statInfo$coef == col_names[which_col],c("feature","Estimate")]
    colnames(method$statInfo) <- c("feature",col_names[which_col])
    rownames(method$statInfo) <- method$statInfo$feature
    
  }
  
  if(col_names[which_col] %in% c("coef_grp2","mle")){# metagenomeSeq has the log-odd values which have the opposite sign
    
    method$statInfo[,col_names[which_col]] <- -method$statInfo[,col_names[which_col]] 
    
  }
  
  # statInfo investigation for DA direction
  direction <- sign(method$statInfo[as.character(DA_features),col_names[which_col]])
  direction <- factor(as.character(direction),levels = c("-1","1"),labels = c("DOWN in Supragingival","UP in Supragingival"))
  # association with genus methabolism
  # meth = methabolism_df[DA_features,"Meth"]
  
  df <- data.frame("Taxa" = HotLoadings.names(ps,format = "last"),row.names = rownames(otu_table(ps)))
  
  df$DA_features <- factor("non-DA",levels = c("non-DA","DOWN in Supragingival","UP in Supragingival"))
  df[as.character(DA_features),"DA_features"] <- direction
  df$Methabolism <- methabolism_df$Meth
  
  return(df)
}
```

DA features (adjusted p-value < 0.1) are extracted from the methods' list. Songbird failed due to the low sample size with the following error: "ValueError: initial_value must have a shape specified: Tensor("random_normal:0", shape=(2, ?), dtype=float32)"

```{r}
DA_methods <- lapply(eval_methods,function(method) 
  extract_statistics(method = method,
                     col_names = col_names,
                     alpha = 0.1,percent = 5,
                     methabolism_df = methabolism_df,ps = ps))
```

A different approach consists in choosig a threshold for taxa selection based on the ordered list of p-values or loadings (mixMC).

```{r}
extract_statistics_threshold <- function(method, col_names = col_names, percent = 30, methabolism_df = methabolism_df, ps){
  which_col <- which(col_names %in% colnames(method$statInfo))
  
  ### LIST OF P-VALUES, DIFFERENTIALS OR LOADINGS
  
  if("pValMat" %in% names(method)){# for most methods
    
    # pval list extraction
    value_list <- data.frame(method$pValMat[,1])
    if(is.null(rownames(method$statInfo))) # Naming the statInfo matrix
      rownames(method$statInfo) <- rownames(method$pValMat)
    
  } else if(col_names[which_col] == "grp.T.grp2."){# songbird
    
    value_list <- data.frame(-abs(method$statInfo[,"grp.T.grp2."]),row.names = method$statInfo[,"featureid"])
    
  } else if(col_names[which_col] == "value.var"){# mixMC
    
    value_list <- data.frame(-abs(method$statInfo[,"value.var"]),row.names = method$statInfo[,"taxa"])
    method$statInfo$value.var <- -method$statInfo$value.var 
    
  } 
  
  top_percent <- order(value_list)[1:round(ntaxa(ps)*percent/100)]
  DA_features <- rownames(value_list)[top_percent]
  
  ### STATINFO EXAMINATIONS
  
  if(col_names[which_col] == "coef"){# corncob methods have melted statInfo matrix
    
    which_col <- which(col_names %in% method$statInfo$coef)
    method$statInfo <- method$statInfo[method$statInfo$coef == col_names[which_col],c("feature","Estimate")]
    colnames(method$statInfo) <- c("feature",col_names[which_col])
    rownames(method$statInfo) <- method$statInfo$feature
    
  }
  
  if(col_names[which_col] %in% c("coef_grp2","mle")){# metagenomeSeq has the log-odd values which have the opposite sign. scde has a different reference
    
    method$statInfo[,col_names[which_col]] <- -method$statInfo[,col_names[which_col]] 
    
  }
  
  ### Investigation for DA direction
  direction <- sign(method$statInfo[DA_features,col_names[which_col]])
  direction <- factor(as.character(direction),levels = c("-1","1"),labels = c("DOWN in Supragingival","UP in Supragingival"))
  # association with genus methabolism
  # meth = methabolism_df[DA_features,"Meth"]
  
  df <- data.frame("Taxa" = HotLoadings.names(ps,format = "last"),row.names = rownames(otu_table(ps)))
  
  df$DA_features <- factor("non-DA",levels = c("non-DA","DOWN in Supragingival","UP in Supragingival"))
  df <- df[DA_features,]
  df[,"DA_features"] <- direction
  df$Methabolism <- methabolism_df[DA_features,"Meth"]
  
  return(df)
}
```

```{r}
extract_statistics_threshold_enrich <- function(method, col_names = col_names, percent = 30, methabolism_df = methabolism_df, ps){
  which_col <- which(col_names %in% colnames(method$statInfo))
  
  ### LIST OF P-VALUES, DIFFERENTIALS OR LOADINGS
  
  if("pValMat" %in% names(method)){# for most methods
    
    # pval list extraction
    value_list <- data.frame(method$pValMat[,1])
    if(is.null(rownames(method$statInfo))) # Naming the statInfo matrix
      rownames(method$statInfo) <- rownames(method$pValMat)
    
  } else if(col_names[which_col] == "grp.T.grp2."){# songbird
    
    value_list <- data.frame(-abs(method$statInfo[,"grp.T.grp2."]),row.names = method$statInfo[,"featureid"])
    
  } else if(col_names[which_col] == "value.var"){# mixMC
    
    value_list <- data.frame(-abs(method$statInfo[,"value.var"]),row.names = method$statInfo[,"taxa"])
    method$statInfo$value.var <- -method$statInfo$value.var
    
  } 
  
  top_percent <- order(value_list)[1:round(ntaxa(ps)*percent/100)]
  DA_features <- rownames(value_list)[top_percent]
  
  ### STATINFO EXAMINATIONS
  
  if(col_names[which_col] == "coef"){# corncob methods have melted statInfo matrix
    
    which_col <- which(col_names %in% method$statInfo$coef)
    method$statInfo <- method$statInfo[method$statInfo$coef == col_names[which_col],c("feature","Estimate")]
    colnames(method$statInfo) <- c("feature",col_names[which_col])
    rownames(method$statInfo) <- method$statInfo$feature
    
  }
  
  if(col_names[which_col] %in% c("coef_grp2","mle")){# metagenomeSeq has the log-odd values which have the opposite sign. scde has a different reference
    
    method$statInfo[,col_names[which_col]] <- -method$statInfo[,col_names[which_col]] 
    
  }
  
  ### Investigation for DA direction
  direction <- sign(method$statInfo[DA_features,col_names[which_col]])
  direction <- factor(as.character(direction),levels = c("-1","1"),labels = c("DOWN in Supragingival","UP in Supragingival"))
  # association with genus methabolism
  # meth = methabolism_df[DA_features,"Meth"]
  
  df <- data.frame("Taxa" = HotLoadings.names(ps,format = "last"),row.names = rownames(otu_table(ps)))
  
  df$DA_features <- factor("non-DA",levels = c("non-DA","DOWN in Supragingival","UP in Supragingival"))
  #df <- df[DA_features,]
  df[DA_features,"DA_features"] <- direction
  df$Methabolism <- methabolism_df[,"Meth"]
  
  return(df)
}
```

```{r, message=FALSE, warning=FALSE}
eval_methods_thresholds <- eval_methods
## mixMC returns only discriminant features which are useful for DA analysis
## Forcing mixMC to choose more taxa (at least 100) will allow to use a thresholds to extract
## the most discriminant ones (as for songbird).
mixMC <- mixMCmodel(ps,grid.keepX = c(400,500))
eval_methods_thresholds$mixMC <- mixMC
DA_methods_threshold <- lapply(eval_methods_thresholds,function(method) 
  extract_statistics_threshold(method = method,
                     col_names = col_names,percent = 5,
                     methabolism_df = methabolism_df,ps = ps))
```

```{r}
extract_statistics_threshold_enrich <- function(method, col_names = col_names, percent = 30, methabolism_df = methabolism_df, ps){
  which_col <- which(col_names %in% colnames(method$statInfo))
  
  ### LIST OF P-VALUES, DIFFERENTIALS OR LOADINGS
  
  if("pValMat" %in% names(method)){# for most methods
    
    # pval list extraction
    value_list <- data.frame(method$pValMat[,1])
    if(is.null(rownames(method$statInfo))) # Naming the statInfo matrix
      rownames(method$statInfo) <- rownames(method$pValMat)
    
  } else if(col_names[which_col] == "grp.T.grp2."){# songbird
    
    value_list <- data.frame(-abs(method$statInfo[,"grp.T.grp2."]),row.names = method$statInfo[,"featureid"])
    
  } else if(col_names[which_col] == "value.var"){# mixMC
    
    value_list <- data.frame(-abs(method$statInfo[,"value.var"]),row.names = method$statInfo[,"taxa"])
    
  } 
  
  top_percent <- order(value_list)[1:round(ntaxa(ps)*percent/100)]
  DA_features <- rownames(value_list)[top_percent]
  
  ### STATINFO EXAMINATIONS
  
  if(col_names[which_col] == "coef"){# corncob methods have melted statInfo matrix
    
    which_col <- which(col_names %in% method$statInfo$coef)
    method$statInfo <- method$statInfo[method$statInfo$coef == col_names[which_col],c("feature","Estimate")]
    colnames(method$statInfo) <- c("feature",col_names[which_col])
    rownames(method$statInfo) <- method$statInfo$feature
    
  }
  
  if(col_names[which_col] %in% c("coef_grp2","mle")){# metagenomeSeq has the log-odd values which have the opposite sign. scde has a different reference
    
    method$statInfo[,col_names[which_col]] <- -method$statInfo[,col_names[which_col]] 
    
  }
  
  ### Investigation for DA direction
  direction <- sign(method$statInfo[DA_features,col_names[which_col]])
  direction <- factor(as.character(direction),levels = c("-1","1"),labels = c("DOWN in Supragingival","UP in Supragingival"))
  # association with genus methabolism
  # meth = methabolism_df[DA_features,"Meth"]
  
  df <- data.frame("Taxa" = HotLoadings.names(ps,format = "last"),row.names = rownames(otu_table(ps)))
  
  df$DA_features <- factor("non-DA",levels = c("non-DA","DOWN in Supragingival","UP in Supragingival"))
  #df <- df[DA_features,]
  df[DA_features,"DA_features"] <- direction
  df$Methabolism <- methabolism_df[,"Meth"]
  
  return(df)
}
```

# Contingency tables

Contingency table are computed and graphically represented using Odds ratios and log odds ratios.

```{r}
create_contingency_tables <- function(method){
  ntaxa = nrow(method)
  method$Anaerobic <- "not Anaerobic"
  method$Anaerobic[method$Methabolism == "Anaerobic"] <- "Anaerobic"
  
  method$Aerobic <- "not Aerobic"
  method$Aerobic[method$Methabolism == "Aerobic"] <- "Aerobic"
  
  method$FAnaerobic <- "not F Anaerobic"
  method$FAnaerobic[method$Methabolism == "F Anaerobic"] <- "F Anaerobic"
  
  method$Unassigned  <- "Assigned"
  method$Unassigned [is.na(method$Methabolism)] <- "Unassigned"
  
  method$DOWN <- "not DOWN"
  method$DOWN[method$DA_features == "DOWN in Supragingival"] <- "DOWN in Supragingival"
  method$DOWN <- factor(method$DOWN, levels = c("DOWN in Supragingival","not DOWN"), ordered = TRUE)
  
  method$UP <- "not UP"
  method$UP[method$DA_features == "UP in Supragingival"] <- "UP in Supragingival"
  method$UP <- factor(method$UP, levels = c("UP in Supragingival","not UP"), ordered = TRUE)
  t_list <- 
  list(t_anaerobic_DOWN = table(method$DOWN,method$Anaerobic),
       t_aerobic_DOWN = table(method$DOWN,method$Aerobic),
       t_fanaerobic_DOWN = table(method$DOWN,method$FAnaerobic),
       t_unassigned_DOWN  = table(method$DOWN,method$Unassigned),
       t_anaerobic_UP = table(method$UP,method$Anaerobic),
       t_aerobic_UP = table(method$UP,method$Aerobic),
       t_fanaerobic_UP = table(method$UP,method$FAnaerobic),
       t_unassigned_UP  = table(method$UP,method$Unassigned))
  
  t_list <- lapply(t_list,function(tab){
    if(ncol(tab) == 1){
      t <- as.table(matrix(c(0,0,0,0),nrow = 2,ncol = 2))
      rownames(t) <- rownames(tab)
      if(grepl(pattern = "not",x = colnames(tab))){
        c_names = c(gsub(colnames(tab),pattern = "not ",replacement = ""),colnames(tab))
      } else c_names = c(colnames(tab),paste0("not ",colnames(tab)))
      colnames(t) <- c_names
      t[,colnames(tab)] <- tab
      tab <- t
    } 
    return(tab)
  })
  
  f_list <- lapply(t_list,function(tab){
    # q = tab[1,1] # Number of DA taxa, in the class of interest
    # k = sum(tab[1,]) # Number DA taxa
    # m = sum(tab[,1]) # Number of taxa in the class of interest
    # n = sum(tab[,2]) # Number of taxa not in the class of interest
    # if(q > 0){
    #   p_value = phyper(q = q-1, m = m, k = k, n = n,lower.tail = FALSE) 
    # } else p_value = 1
    f <- fisher.test(tab,alternative = "g") # Haldane-Anscombe correction
    # f$p.value = p_value
    tabm <- tab+0.5
    f$estimate <- (tabm[1,1]*tabm[2,2])/(tabm[1,2]*tabm[2,1])
    return(f)
  })
  names(f_list) <- sapply(names(f_list),gsub,pattern = "t_",replacement = "f_")
  
  ci_list <- lapply(t_list,function(tab){
    1.96*sqrt(sum(1/(tab + 1)))
  })
  names(ci_list) <- sapply(names(ci_list),gsub,pattern = "t_",replacement = "ci_")
  
  s_list <- lapply(t_list,function(tab){
    (sum(c(tab[1,1],tab[2,2])) - mean(c(tab[1,2],tab[2,1])) - sd(c(tab[1,1],tab[2,2])))/ntaxa
  })
  names(s_list) <- sapply(names(s_list),gsub,pattern = "t_",replacement = "s_")
  
  
  
  return(c(t_list,f_list,ci_list,s_list))
}

create_contingency_tables_omogeneity <- function(method){
  
  ntaxa = nrow(method)
  method <- method %>% filter(Methabolism %in% c("Aerobic","Anaerobic"))
  method$Methabolism <- factor(method$Methabolism,levels = c("Aerobic","Anaerobic"))
  
  tab = table(method$DA_features,method$Methabolism)
  
  if(ncol(tab) == 1){
    t <- as.table(matrix(c(0,0,0,0),nrow = 2,ncol = 2))
    rownames(t) <- rownames(tab)
    colnames(t) <- colnames(tab)
    t[,colnames(tab)] <- tab
    tab <- t
  }
  
  tab <- tab[c("UP in Supragingival","DOWN in Supragingival"),]
  
  f <- fisher.test(tab,alternative = "g") # Haldane-Anscombe correction
  tabm <- tab+0.5
  f$estimate <- (tabm[1,1]*tabm[2,2])/(tabm[1,2]*tabm[2,1])
  s <- (sum(c(tab[1,1],tab[2,2])) - mean(c(tab[1,2],tab[2,1])) - sd(c(tab[1,1],tab[2,2])))/ntaxa
  
  s <- (sum(c(tab[1,1],tab[2,2])) - sum(c(tab[1,2],tab[2,1])))/ntaxa
  x <- chisq.test(tab+1,correct = TRUE)
  
  ci <- 1.96*sqrt(sum(1/(tab+1)))
  
  return(list(t = tab, f = f, x = x, ci = ci, s = s))
}

plot_contingency_table <- function(data_to_plot, methabolism, method_name, direction, omogeneity = FALSE){
  if(!omogeneity)
    data_to_plot <- data_to_plot[data_to_plot$variable %in% c(methabolism,paste0("not ",methabolism)) &  grepl(pattern = direction,data_to_plot$Direction),]
  
  data_to_plot = data_to_plot %>% filter(Method == method_name)
  
  ggplot(data = data_to_plot,mapping = aes(x = variable, y = Direction, fill = value)) + 
    geom_tile(width = 0.8, height = 0.8) +
    geom_label(aes(label = value), fill = "white") +
    coord_fixed() + 
    theme_minimal() +
    scale_x_discrete(position = "top") +
    ylab(label = element_blank()) +
    theme(legend.position = "none",
          axis.title.x = element_blank()) +
    ggtitle(label = element_blank(),subtitle = method_name)
}
```

```{r, warning=FALSE,message=FALSE}
contingency_table_list <- lapply(DA_methods,create_contingency_tables)

# Extract counts
contingency_table_DOWN_df <- ldply(contingency_table_list,.fun = function(method){
  cbind(method$t_aerobic_DOWN,
        method$t_anaerobic_DOWN,
        method$t_fanaerobic_DOWN,
        method$t_unassigned_DOWN)
},.id = "Method")

contingency_table_UP_df <- ldply(contingency_table_list,.fun = function(method){
  cbind(method$t_aerobic_UP,
        method$t_anaerobic_UP,
        method$t_fanaerobic_UP,
        method$t_unassigned_UP)
},.id = "Method")

contingency_table_DOWN_df$Direction <- as.factor(factor(rownames(contingency_table_list$mixMC$t_anaerobic_DOWN)))
contingency_table_UP_df$Direction <- as.factor(factor(rownames(contingency_table_list$mixMC$t_anaerobic_UP)))

contingency_table_DOWN_melted <- melt(contingency_table_DOWN_df)
contingency_table_UP_melted <- melt(contingency_table_UP_df)

contingency_table_melted <- rbind(contingency_table_DOWN_melted,
                                  contingency_table_UP_melted)

contingency_table_melted$Direction <- factor(contingency_table_melted$Direction, levels = c("not DOWN","DOWN in Supragingival","not UP","UP in Supragingival"), ordered = TRUE)

odds_ratio_df_calculation <- function(method, direction, methabolism){
  vec <- paste(c("ci","f"),methabolism,direction,sep = "_")
  CI = method[[vec[1]]]
  f_CI = method[[vec[2]]]$conf.int[1:2]
  f_est = method[[vec[2]]]$estimate
  f_pvalue = method[[vec[2]]]$p.value
  return(data.frame("CI_logodds" = CI,
                    "LowerBound" = f_CI[1],
                    "UpperBound" = f_CI[2],
                    "Odds ratio" = f_est,
                    "p-value" = f_pvalue,
                    "Methabolism" = as.character(methabolism),
                    "Direction" = direction))
}

Aerobic_UP_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "UP",methabolism = "aerobic"),.id = "Method")
Aerobic_DOWN_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "DOWN",methabolism = "aerobic"),.id = "Method")

Anaerobic_UP_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "UP",methabolism = "anaerobic"),.id = "Method")
Anaerobic_DOWN_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "DOWN",methabolism = "anaerobic"),.id = "Method")

FAnaerobic_UP_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "UP",methabolism = "fanaerobic"),.id = "Method")
FAnaerobic_DOWN_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "DOWN",methabolism = "fanaerobic"),.id = "Method")

Unassigned_UP_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "UP",methabolism = "unassigned"),.id = "Method")
Unassigned_DOWN_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) odds_ratio_df_calculation(method = method,direction = "DOWN",methabolism = "unassigned"),.id = "Method")

odds_ratio_df <- rbind(Aerobic_DOWN_odds_ratio_df,
                       Aerobic_UP_odds_ratio_df,
                       Anaerobic_DOWN_odds_ratio_df,
                       Anaerobic_UP_odds_ratio_df,
                       FAnaerobic_DOWN_odds_ratio_df,
                       FAnaerobic_UP_odds_ratio_df,
                       Unassigned_DOWN_odds_ratio_df,
                       Unassigned_UP_odds_ratio_df)

odds_ratio_df$Methabolism <- factor(odds_ratio_df$Methabolism, levels = c("aerobic","anaerobic","fanaerobic","unassigned"), labels = c("Aerobic","Anaerobic","F Anaerobic","Unassigned"))

odds_ratio_df$Direction <- factor(odds_ratio_df$Direction, levels = c("DOWN","UP"), labels = c("DOWN in Supragingival","UP in Supragingival"))
```

```{r, warning=FALSE, message=FALSE}
contingency_table_list_omogeneity <- lapply(DA_methods_threshold,create_contingency_tables_omogeneity)

contingency_table_omogeneity <- ldply(contingency_table_list_omogeneity,.fun = function(method){ method$t },.id = "Method")

contingency_table_omogeneity$Direction <- as.factor(factor(rownames(contingency_table_list_omogeneity$mixMC$t)))

contingency_table_omogeneity_melted <- melt(contingency_table_omogeneity)
```

Many ASVs are not characterized at the genus level:

## Adjusted p-value < 0.1

```{r}
lapply(DA_methods,function(method) {
  summary(method[method$DA_features != "non-DA",]$Methabolism)
})
```

Considering the following things:
<ul>
  <li>the reference group is the **subgingival**;</li>
  <li>UP-regulated taxa are more abundant in **supragingival** rather than **subgingival** plaque;</li>
  <li>DOWN-regulated taxa are more abundant in **subgingival** rather than **supragingival** plaque;</li>
  <li>contingency matrix has the number of DOWN-regulated taxa as first row;</li>
  <li>an odds-ratio greater than 1 (log-odds ratio > 0) means that the odd between subgingival and supragingival taxa is greater in the considered methabolic class rather than the odd in the other class. </li>
  <li>we expect positive log-odds ratio for the anaerobic methabolism. This is caused by the presence of more anaerobic bacteria in subgingival plaque rather than supragingival plaque.</li>
</ul>

```{r, fig.height=8, fig.width=18}
plot_list_Aerobic_UP <- lapply(names(contingency_table_list),function(method) plot_contingency_table(data_to_plot = contingency_table_melted, methabolism = "Aerobic",method_name = method, direction = "UP"))
plot_list_Aerobic_DOWN <- lapply(names(contingency_table_list),function(method) plot_contingency_table(data_to_plot = contingency_table_melted,methabolism = "Aerobic",method_name = method, direction = "DOWN"))

plot_list_Anaerobic_UP <- lapply(names(contingency_table_list),function(method) plot_contingency_table(data_to_plot = contingency_table_melted,methabolism = "Anaerobic",method_name = method, direction = "UP"))
plot_list_Anaerobic_DOWN <- lapply(names(contingency_table_list),function(method) plot_contingency_table(data_to_plot = contingency_table_melted,methabolism = "Anaerobic",method_name = method, direction = "DOWN"))

plot_list_FAnaerobic_UP <- lapply(names(contingency_table_list),function(method) plot_contingency_table(data_to_plot = contingency_table_melted,methabolism = "F Anaerobic",method_name = method, direction = "UP"))
plot_list_FAnaerobic_DOWN <- lapply(names(contingency_table_list),function(method) plot_contingency_table(data_to_plot = contingency_table_melted,methabolism = "F Anaerobic",method_name = method, direction = "DOWN"))

plot_grid(plotlist = plot_list_Aerobic_UP,nrow = 4, ncol = 5)
plot_grid(plotlist = plot_list_Aerobic_DOWN,nrow = 4, ncol = 5)
plot_grid(plotlist = plot_list_Anaerobic_UP,nrow = 4, ncol = 5)
plot_grid(plotlist = plot_list_Anaerobic_DOWN,nrow = 5, ncol = 5)
plot_grid(plotlist = plot_list_FAnaerobic_UP,nrow = 4, ncol = 5)
plot_grid(plotlist = plot_list_FAnaerobic_DOWN,nrow = 4, ncol = 5)
```

## Threshold fixed for all methods percent = 5%

```{r}
lapply(DA_methods_threshold,function(method) {
  summary(method[method$DA_features != "non-DA",]$Methabolism)
})
```

Here an example for the UP vs DOWN, Aerobic vs Anaerobic top % taxa contingency tables. On these tables a $\chi^2$ test of omogeneity, Fisher exact test and log-odds ratio are computed. The null hypothesys in this case is that UP and DOWN regulated taxa are omogeneously distributed between Aerobic and Anaerobic taxa

```{r, fig.height=8, fig.width=18}
plot_list_omogeneity <- lapply(names(contingency_table_list_omogeneity),function(method) plot_contingency_table(data_to_plot = contingency_table_omogeneity_melted,method_name = method,omogeneity = TRUE))

plot_grid(plotlist = plot_list_omogeneity,nrow = 4, ncol = 5)
```

## Summary figure

To summarize all the contingency tables, here is presented a barplot where each bar represents the value on the contingency tables in position (1,1).

An hypergeometric test is performed to inspect DA taxa enrichment in the many classes of interest.

Ordered by the number of DA taxa.

```{r, fig.height= 10,fig.width=10}
data_to_plot <- contingency_table_melted %>% 
  filter(Direction %in% c("DOWN in Supragingival","UP in Supragingival")) %>%
  filter(variable %in% c("Aerobic","Anaerobic","F Anaerobic","Unassigned"))

odds_ratio_to_plot <- odds_ratio_df[order(odds_ratio_df[,"Direction"]),]
data_to_plot$p_value <- odds_ratio_to_plot$p.value

ord_fig_a <- order(ddply(data_to_plot,.variables = ~ Method, function(met) sum(abs(met[met$Direction %in% c("DOWN in Supragingival","UP in Supragingival") & met$variable %in% c("Aerobic","Anaerobic","F Anaerobic"),"value"])))$V1)

# Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

fig_a <- ggplot(data = data_to_plot[data_to_plot$variable != "Unassigned",], aes(x = Method,y = ifelse(Direction == "UP in Supragingival",value,-value),fill = variable)) +
  geom_col(position = "dodge",alpha = 0.5) + 
  geom_text(aes(x = Method, y = ifelse(Direction == "UP in Supragingival",value+5,-value-5), label = ifelse(p_value>0.1,"",ifelse(p_value>0.05,"•",ifelse(p_value>0.01,"*",ifelse(p_value>0.001,"**","***")))), color = variable), position = position_dodge(width = 1)) + 
  geom_hline(aes(yintercept = 0), color = "black") +
  scale_y_continuous(breaks = seq(-60,60,length.out = 13),labels = c(seq(from = 60,to = 0,length.out = 7),seq(from = 10,to = 60, length.out = 6)),sec.axis = sec_axis(~./30, name = "DA direction",breaks = c(-1,1),labels = c("DOWN in Supragingival","UP in Supragingival"))) +
  scale_x_discrete(limits = levels(data_to_plot$Method)[ord_fig_a]) +
  theme_minimal() +
  ylab("Number of DA taxa") +
  scale_fill_manual(values = c(RColorBrewer::brewer.pal(3,"Set1"),"darkgrey")) +
  scale_color_manual(values = c(RColorBrewer::brewer.pal(3,"Set1"),"darkgrey")) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        axis.text.y.right = element_text(angle = -90,hjust = 0.5),
        legend.position = "bottom") +
  guides(fill = guide_legend(title = "Metabolism",override.aes = aes(label = "")),color = "none") +
  ggtitle("Number of DA taxa, Supragingival vs Subgingival Plaque",subtitle = "Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘•’ 0.1 ‘ ’ 1")
fig_a
```

## Log Odds ratio

Computed using contingency tables

```{r, fig.width=6, fig.height=7}

ord_df <- ddply(odds_ratio_df,.variables = ~ Direction + Method, function(met) met[met$Methabolism == "Aerobic","Odds.ratio"]-met[met$Methabolism == "Anaerobic","Odds.ratio"])

ord <- order(ord_df[ord_df$Direction == "DOWN in Supragingival","V1"])

ggplot(data = odds_ratio_df %>% filter(Direction == "DOWN in Supragingival" & Methabolism %in% c("Aerobic","Anaerobic")) ,mapping = aes(x = Method, y = log(Odds.ratio), color = Methabolism)) + 
  theme_grey() +
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5)) +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(mapping = aes(ymin = log(Odds.ratio)-CI_logodds, ymax = log(Odds.ratio)+CI_logodds),width = 0.5) +
  scale_x_discrete(limits = levels(odds_ratio_df$Method)[ord]) +
  ylab("Log Odds Ratio") +
  #facet_zoom(ylim = c(-4,4),split = TRUE) +
  geom_hline(mapping = aes(yintercept = 0), color = "black", lty = 2) +
  ggtitle(label = "Log Odds Ratio for DOWN regulated taxa", subtitle = "Aerobic and Anaerobic focus")

ord <- order(ord_df[ord_df$Direction == "UP in Supragingival","V1"])

ggplot(data = odds_ratio_df %>% filter(Direction == "UP in Supragingival" & Methabolism %in% c("Aerobic","Anaerobic")) ,mapping = aes(x = Method, y = log(Odds.ratio), color = Methabolism)) + 
  theme_grey() +
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5)) +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(mapping = aes(ymin = log(Odds.ratio)-CI_logodds, ymax = log(Odds.ratio)+CI_logodds),width = 0.5) +
  scale_x_discrete(limits = levels(odds_ratio_df$Method)[ord]) +
  ylab("Log Odds Ratio") +
  #facet_zoom(ylim = c(-4,4),split = TRUE) +
  geom_hline(mapping = aes(yintercept = 0), color = "black", lty = 2) +
  ggtitle(label = "Log Odds Ratio for UP regulated taxa", subtitle = "Aerobic and Anaerobic focus")
```

```{r, fig.height=7, fig.width=8}
fig_lor_WMS <- ggplot(data = odds_ratio_df %>% filter(Methabolism %in% c("Aerobic","Anaerobic")) ,mapping = aes(x = Method, y = log(Odds.ratio), color = Methabolism)) + 
  facet_wrap(~ Direction, labeller = labeller(Direction = c(`DOWN in Supragingival` = "DOWN in Supragingival - Anaerobic",`UP in Supragingival` = "UP in Supragingival - Aerobic"))) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90,hjust = 1, vjust = 0.5),
        legend.position = "bottom") +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(mapping = aes(ymin = log(Odds.ratio)-CI_logodds, ymax = log(Odds.ratio)+CI_logodds),width = 0.5) +
  scale_x_discrete(limits = levels(odds_ratio_df$Method)[ord_fig_a]) +
  ylab("Log Odds Ratio") + labs(color = "Metabolism") +
  #facet_zoom(ylim = c(-4,4),split = TRUE) +
  geom_hline(mapping = aes(yintercept = 0), color = "black", lty = 2) +
  ggtitle(label = "Log Odds Ratio, Supragingival vs Subgingival Plaque, WMS", subtitle = "Aerobic and Anaerobic taxa focus")
fig_lor_WMS

```

```{r, include=FALSE, fig.width=8, fig.height=12}

svg(filename = "../../fig_lor.svg",width = 8,height = 12)
plot_grid(plotlist = list(fig_lor_16S,fig_lor_WMS),labels = "auto",ncol = 1)
dev.off()
```


## Odds ratio

```{r,fig.width=10, fig.height=7}

ord_df <- ddply(odds_ratio_df,.variables = ~ Direction + Method, function(met) met[met$Methabolism == "Aerobic","Odds.ratio"]-met[met$Methabolism == "Anaerobic","Odds.ratio"])

ord <- order(ord_df[ord_df$Direction == "DOWN in Supragingival","V1"])

ggplot(data = odds_ratio_df %>% filter(Direction == "DOWN in Supragingival" & Methabolism %in% c("Aerobic","Anaerobic")) ,mapping = aes(x = Method, y = Odds.ratio, color = Methabolism)) + 
  theme_grey() +
  theme(axis.text.x = element_text(angle = 90,hjust = 1),
        axis.text.y.right = element_text(angle = 90,hjust = 0.5),
        axis.title.y.right = element_text(angle = 90),
        axis.title.x = element_text(angle = 180)) +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(mapping = aes(ymin = LowerBound, ymax = UpperBound),width = 0.5) +
  scale_x_discrete(limits = levels(odds_ratio_df$Method)[ord]) +
  scale_y_continuous(position = "right") +
  ylab("Odds Ratio") +
  facet_zoom(ylim = c(0,5),split = TRUE) +
  geom_hline(mapping = aes(yintercept = 1), color = "black", lty = 2) +
  ggtitle(label = "Odds Ratio for UP regulated taxa", subtitle = "Aerobic and Anaerobic focus")

ord <- order(ord_df[ord_df$Direction == "UP in Supragingival","V1"])

ggplot(data = odds_ratio_df %>% filter(Direction == "UP in Supragingival" & Methabolism %in% c("Aerobic","Anaerobic")) ,mapping = aes(x = Method, y = Odds.ratio, color = Methabolism)) + 
  theme_grey() +
  theme(axis.text.x = element_text(angle = 90,hjust = 1),
        axis.text.y.right = element_text(angle = 90,hjust = 0.5),
        axis.title.y.right = element_text(angle = 90),
        axis.title.x = element_text(angle = 180)) +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(mapping = aes(ymin = LowerBound, ymax = UpperBound),width = 0.5) +
  scale_x_discrete(limits = levels(odds_ratio_df$Method)[ord]) +
  scale_y_continuous(position = "right") +
  ylab("Odds Ratio") +
  facet_zoom(ylim = c(0,5),split = TRUE) +
  geom_hline(mapping = aes(yintercept = 1), color = "black", lty = 2) +
  ggtitle(label = "Odds Ratio for UP regulated taxa", subtitle = "Aerobic and Anaerobic focus")
```

## Mutual findings plot

Each method finds some DA taxa. In this plot we inspect wether they share the same findings or not.

```{r}
DA_methods_df <- ldply(DA_methods,function(met) met %>% filter(DA_features != "non-DA"),.id = "Method")
DA_methods_df$Compositional <- factor(DA_methods_df$Method,
                                      levels = c("mixMC",
                                                "seurat_wilcoxon",
                                                "scde",
                                                "MAST",
                                                "corncob_LRT",
                                                "corncob_wald",
                                                "ALDEx2",
                                                "mgsZig_CSS",
                                                "DESeq2_TMM",
                                                "DESeq2_poscounts_zinbwave",
                                                "DESeq2_poscounts",
                                                "limma_voom_TMM_zinbwave",
                                                "limma_voom_TMM",
                                                "edgeR_poscounts_standard",
                                                "edgeR_TMM_zinbwave",
                                                "edgeR_TMM_standard",
                                                "edgeR_TMM_robustDisp",
                                                "songbird"),
                                      labels = c("Compositional","Non-Compositional","Non-Compositional","Non-Compositional","Non-Compositional","Non-Compositional",
                                                 "Compositional","Non-Compositional","Non-Compositional","Non-Compositional","Non-Compositional","Non-Compositional",
                                                 "Non-Compositional","Non-Compositional","Non-Compositional","Non-Compositional","Non-Compositional","Compositional"))

DA_methods_df$Tech <- factor(DA_methods_df$Method,
                                      levels = c("mixMC",
                                                "seurat_wilcoxon",
                                                "scde",
                                                "MAST",
                                                "corncob_LRT",
                                                "corncob_wald",
                                                "ALDEx2",
                                                "mgsZig_CSS",
                                                "DESeq2_TMM",
                                                "DESeq2_poscounts_zinbwave",
                                                "DESeq2_poscounts",
                                                "limma_voom_TMM_zinbwave",
                                                "limma_voom_TMM",
                                                "edgeR_poscounts_standard",
                                                "edgeR_TMM_zinbwave",
                                                "edgeR_TMM_standard",
                                                "edgeR_TMM_robustDisp",
                                                "songbird"),
                                      labels = c("metagenomics","sc-RNAseq","sc-RNAseq","sc-RNAseq","metagenomics","metagenomics",
                                                 "metagenomics","metagenomics","RNAseq","sc-RNAseq","RNAseq","sc-RNAseq",
                                                 "RNAseq","RNAseq","sc-RNAseq","RNAseq","RNAseq","metagenomics"))

length(unique(DA_methods_df$Taxa))

taxa_by_method <- ddply(DA_methods_df,.variables = ~ Taxa, .fun = function(taxa) nrow(taxa))
colnames(taxa_by_method) <- c("Taxa","Number of methods")
ggplot(taxa_by_method,aes(x = `Number of methods`)) +
  geom_bar(width = 0.9) +
  theme_minimal() + 
  scale_x_continuous(breaks = 1:18) +
  xlab("Number of Methods") + 
  ggtitle("# of methods per taxa")

taxa_by_method <- ddply(DA_methods_df,.variables = ~ Taxa + Compositional, .fun = function(taxa) nrow(taxa))
colnames(taxa_by_method) <- c("Taxa","Compositional","Number of methods")
ggplot(taxa_by_method,aes(x = `Number of methods`)) +
  geom_bar(width = 0.9) + 
  facet_grid(~ Compositional,scales = "free_x") +
  theme_minimal() +
  xlab("Number of Methods") + 
  ggtitle("# of methods per taxa",subtitle = "By compositionality")

taxa_by_method <- ddply(DA_methods_df,.variables = ~ Taxa + Tech, .fun = function(taxa) nrow(taxa))
colnames(taxa_by_method) <- c("Taxa","Tech","Number of methods")
ggplot(taxa_by_method,aes(x = `Number of methods`)) +
  geom_bar(width = 0.9) + 
  facet_grid(~ Tech,scales = "free_x") +
  theme_minimal() +
  xlab("Number of Methods") + 
  ggtitle("# of methods per taxa",subtitle = "By application")

taxa_by_methabolism <- ddply(DA_methods_df,.variables = ~ Taxa + Methabolism, .fun = function(taxa) nrow(taxa))
colnames(taxa_by_methabolism) <- c("Taxa","Methabolism","Number of methods")
ggplot(taxa_by_methabolism,aes(x = `Number of methods`)) +
  geom_bar(width = 0.9) + 
  facet_grid(~ Methabolism,scales = "free_x") +
  theme_minimal() +
  xlab("Number of Methods") + 
  ggtitle("# of methods per taxa",subtitle = "By methabolism")

length(unique(taxa_by_methabolism[taxa_by_methabolism$Methabolism == "Aerobic","Taxa"]))

```

215 Unique taxa out of 516 are found as DA from one or more method. However only a minority are mutually found by all of them. 

```{r}
representative_methods <- c("mixMC","seurat_wilcoxon","scde",
                            "MAST","corncob_wald","ALDEx2",
                            "mgsZig_CSS","DESeq2_poscounts",
                            "limma_voom_TMM","edgeR_TMM_standard")

mutual_findings <- function(DA_methods_df, min_num = 5, DA_feat, Meth, compositional = c("Compositional","Non-Compositional"), tech = c("metagenomics","sc-RNAseq","RNAseq"), methods = NULL){
  
  DA_methods_df <- DA_methods_df %>% filter(Tech %in% tech & Compositional %in% compositional)
  
  taxa_by_method <- ddply(DA_methods_df,.variables = ~ Taxa, .fun = function(taxa) nrow(taxa))
  taxa_to_plot <- taxa_by_method %>% filter(V1 >= min_num)
  
  df_to_plot <- DA_methods_df %>% filter(Taxa %in% taxa_to_plot$Taxa) %>% filter(Methabolism %in% Meth & DA_features %in% DA_feat)
  
  if(!is.null(methods)){
    df_for_taxa_selection <- df_to_plot %>% filter(Method %in% methods)
    taxa_by_method_for_selection <- ddply(df_for_taxa_selection,.variables = ~ Taxa, .fun = function(taxa) nrow(taxa))
    taxa_selected <- taxa_by_method_for_selection %>% filter(V1 >= min_num)
    df_to_plot <- df_to_plot %>% filter(Taxa %in% unique(as.character(taxa_selected$Taxa)))
  }
  
  df_to_plot$Taxa <- unlist(sapply(as.character(df_to_plot$Taxa),function(x) strsplit(x = x,split = "\\|")[[1]][1]))
  
  y_lab <- ddply(df_to_plot,.variables = ~ Taxa, .fun = function(taxa) nrow(taxa))
  ord_y <- order(y_lab$V1,decreasing = TRUE)
  
  x_lab <- ddply(df_to_plot,.variables = ~ Method, .fun = function(met) nrow(met))
  ord_x <- order(x_lab$V1,decreasing = TRUE)
  
  last_name <- HotLoadings.names(ps,format = "last")
  
  
  ggplot(data = df_to_plot,aes(x = Method, y = Taxa, fill = DA_features)) + 
    geom_tile(width = 0.9, height = 0.9) + 
    facet_grid(~ Methabolism) +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
          legend.position = "bottom",
          axis.title.y.left = element_blank(),
          strip.text.x = element_blank()) +
    scale_y_discrete(limits = y_lab$Taxa[ord_y],position = "left") +
    scale_x_discrete(limits = x_lab$Method[ord_x]) + 
    coord_equal() +
    labs(fill = "DA features") +
    ggtitle(label = paste0("Mutual findings - ",Meth," metabolism Taxa"),subtitle = paste0("Taxa identified in ",min_num," or more ",ifelse(length(compositional)==2,"",paste0(" ", compositional)),ifelse(length(tech)==3,"",paste0(" ",tech)),"representative Methods"))
}
```

### Aerobic

```{r, fig.width=10, fig.height=8}
mf_aerobic <- mutual_findings(DA_methods_df,min_num = 2,DA_feat = c("DOWN in Supragingival","UP in Supragingival"),Meth = "Aerobic",methods = representative_methods)

mf_aerobic

# mutual_findings(DA_methods_df,min_num = 2,DA_feat = c("DOWN in Supragingival","UP in Supragingival"),Meth = "Aerobic",tech = "sc-RNAseq",compositional = "Non-Compositional")
# 
# mutual_findings(DA_methods_df,min_num = 2,DA_feat = c("DOWN in Supragingival","UP in Supragingival"),Meth = "Aerobic",tech = "metagenomics",compositional = "Non-Compositional")
# 
# mutual_findings(DA_methods_df,min_num = 3,DA_feat = c("DOWN in Supragingival","UP in Supragingival"),Meth = "Aerobic",tech = "RNAseq",compositional = "Non-Compositional")
```

### Anaerobic

```{r, fig.width=10, fig.height=8}
mf_anaerobic <- mutual_findings(DA_methods_df,min_num = 4,DA_feat = c("DOWN in Supragingival","UP in Supragingival"),Meth = "Anaerobic", methods = representative_methods)
mf_anaerobic
```

### F Anaerobic

```{r, fig.width=10, fig.height=10}
mf_fanaerobic <- mutual_findings(DA_methods_df,min_num = 3,DA_feat = c("DOWN in Supragingival","UP in Supragingival"),Meth = "F Anaerobic",methods = representative_methods)
mf_fanaerobic
```

# Methods' ranking

Top ranked taxa, using the list of raw p-values, differentials or loadings are extracted from each method. 

Several thresholds are considered: 1%, 2%, ... , 10% top ranked features. 

Contingency table are built and a pseudocount value of 0.1 is added in order to allow odds-ratio and p-values calculations. Indeed, $\chi^2$ test can be computed if marginals are different from 0. Not added for fisher exact test.

## Possible statistics

We can study methods' performances using:
<ul>
<li> $\chi^2$ p-values, which measures the omogeneity between UP and DOWN regulated taxa between Aerobic and Anaerobic findings; </li>
<li> Fisher exact test p-values, which are better since they rely on an exact test; </li>
<li> Odds ratio or log odds ratio for the contingency table, higher values are related to a stronger association between the variables.</li>

```{r, message=FALSE, warning=FALSE}
percents = as.list(seq(1,20,by = 1))

# To produce UPvsDOWN, Aerobic vs Anaerobic
DA_list <- lapply(percents,function(percent){
  lapply(eval_methods_thresholds,function(method) 
  extract_statistics_threshold(method = method,
                     col_names = col_names,percent = percent,
                     methabolism_df = methabolism_df,ps = ps))
})

names(DA_list) <- as.character(seq(1,20,by = 1))

## Statistics in UPvsDOWN, AerobicvsAnaerobic tables

# Chi-square
chisq_test_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    chisq <- as.vector(create_contingency_tables_omogeneity(method)$x$p.value)
    return(data.frame("p.value" = chisq))
  },.id = "Method")
},.id = "Threshold")

ggplot(chisq_test_df,mapping = aes(x = Threshold,y = p.value, color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  ggtitle(label = "Chi-Squared omogeneity test")


# Odds ratio
oddsratio_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    odds.ratio <- as.vector(create_contingency_tables_omogeneity(method)$f$estimate)
    return(data.frame("odds.ratio" = odds.ratio))
  },.id = "Method")
},.id = "Threshold")

ggplot(oddsratio_df,mapping = aes(x = Threshold,y = log(odds.ratio), color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  ggtitle(label = "Log Odds Ratio") +
  ylab("Log(Odds Ratio)")

# F test
ftest_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    p.value <- as.vector(create_contingency_tables_omogeneity(method)$f$p.value)
    return(data.frame("p.value" = p.value))
  },.id = "Method")
},.id = "Threshold")

ggplot(ftest_df,mapping = aes(x = Threshold,y = 1-p.value, color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  ggtitle(label = "Fisher exact test")

# Statistic based on penalized proportion of TP. Based on UPvsDOWN, AerobicvsAnaerobic tables
sstatistic_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    s.value <- as.vector(create_contingency_tables_omogeneity(method)$s)
    return(data.frame("s.value" = s.value))
  },.id = "Method")
},.id = "Threshold")

ggplot(sstatistic_df,mapping = aes(x = Threshold,y = s.value, color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  ggtitle(label = "(TP-FP)/ntaxaDA")
```


```{r message=FALSE, warning=FALSE}
percents = as.list(seq(1,20,by = 1))
# To produce all contingency tables
DA_list <- lapply(percents,function(percent){
  lapply(eval_methods_thresholds,function(method) 
  extract_statistics_threshold_enrich(method = method,
                     col_names = col_names,percent = percent,
                     methabolism_df = methabolism_df,ps = ps))
})

names(DA_list) <- as.character(seq(1,20,by = 1))

## Statistics in specific tables

# TP - FP statistic
TP_FP_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    cont_table_list <- create_contingency_tables(method)
    
    ae_UP <- cont_table_list$t_aerobic_UP[1,1]
    anae_DOWN <- cont_table_list$t_anaerobic_DOWN[1,1]
    TP <- ae_UP + anae_DOWN
    
    ae_DOWN <- cont_table_list$t_aerobic_DOWN[1,1]
    anae_UP <- cont_table_list$t_anaerobic_UP[1,1] 
    FP <- ae_DOWN + anae_UP
    
    statistic <- (TP-FP)
    statistic_normalized <- statistic/(TP+FP)
    return(data.frame("s" = statistic, "s_norm" = statistic_normalized))
  },.id = "Method")
},.id = "Threshold")

fig_b_TP_FP_method <- ggplot(TP_FP_df,mapping = aes(x = Threshold,y = s, color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  theme_minimal() +
  theme(legend.position = "bottom") + 
  guides(color = guide_legend(nrow = 7)) +
  scale_color_manual(values = cols) + 
  labs(y = "Putative TP - Putative FP", x = "Threshold (% of taxa)") +
  ggtitle(label = "Putative TP - Putative FP", subtitle = "Considering Aerobic and Anaerobic taxa")
fig_b_TP_FP_method

fig_b_TP_FP_norm_method <- ggplot(TP_FP_df,mapping = aes(x = Threshold,y = s_norm, color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  theme_minimal() +
  labs(y = "(Putative TP - Putative FP)/(Putative TP + Putative FP)") +
  ggtitle(label = "Normalized Putative TP - Putative FP", subtitle = "Considering Aerobic and Anaerobic taxa")
fig_b_TP_FP_norm_method

TP_FP_AdjPVal_df <- ldply(.data = DA_methods,.fun = function(method){
    cont_table_list <- create_contingency_tables(method)
    
    ae_UP <- cont_table_list$t_aerobic_UP[1,1]
    anae_DOWN <- cont_table_list$t_anaerobic_DOWN[1,1]
    TP <- ae_UP + anae_DOWN
    
    ae_DOWN <- cont_table_list$t_aerobic_DOWN[1,1]
    anae_UP <- cont_table_list$t_anaerobic_UP[1,1] 
    FP <- ae_DOWN + anae_UP
    
    statistic <- (TP-FP)
    statistic_normalized <- statistic/(TP+FP)
    return(data.frame("s" = statistic, "s_norm" = statistic_normalized))
  },.id = "Method")

ggplot(TP_FP_AdjPVal_df,mapping = aes(x = "Supragingival vs Subgingival",y = Method, fill = s)) + 
  geom_tile(width = 0.8, height = 0.8) +
  coord_fixed() +
  theme_minimal() +
  scale_fill_distiller(palette = "RdYlBu") +
  labs(y = "Putative TP - Putative FP") +
  ggtitle(label = "Putative TP - Putative FP", subtitle = "Considering Aerobic and Anaerobic taxa, Adj p-value < 0.1")

ggplot(TP_FP_AdjPVal_df,mapping = aes(x = "Supragingival vs Subgingival",y = Method, fill = s_norm)) + 
  geom_tile(width = 0.8, height = 0.8) +
  coord_fixed() +
  theme_minimal() +
  scale_fill_distiller(palette = "RdYlBu") +
  labs(y = "(Putative TP - Putative FP)/(Putative TP + Putative FP)") +
  ggtitle(label = "Normalized Putative TP - Putative FP", subtitle = "Considering Aerobic and Anaerobic taxa, Adj p-value < 0.1")

# Fisher comination of p-values
combined_enrichment_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    cont_table_list <- create_contingency_tables(method)
    p1 <- cont_table_list$f_aerobic_UP$p.value
    p2 <- cont_table_list$f_anaerobic_DOWN$p.value
    
    statistic <- -2*(log(p1)+log(p2))
    p.value = pchisq(q = statistic,df = 4,lower.tail = FALSE)
    
    return(data.frame("p.value" = p.value))
  },.id = "Method")
},.id = "Threshold")

ggplot(combined_enrichment_df,mapping = aes(x = Threshold,y = -log10(p.value), color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  ggtitle(label = "Fisher combined p.values",subtitle = "Aerobic and Anaerobic enrichment test p.values")

# Chi-squared on 3x3 matrix. UP,DOWN,nonDA and Aerobic,Anaerobic,Other table
chi3_3_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    cont_table_list <- create_contingency_tables(method)
    m <- matrix(data = 0L,ncol = 3, nrow = 3)
    m[1,1] <- cont_table_list$t_aerobic_UP[1,1]
    m[1,2] <- cont_table_list$t_anaerobic_UP[1,1]
    m[1,3] <- cont_table_list$t_fanaerobic_UP[1,1] + 
      cont_table_list$t_unassigned_UP[1,2]
      
    m[2,1] <- cont_table_list$t_aerobic_DOWN[1,1]
    m[2,2] <- cont_table_list$t_anaerobic_DOWN[1,1]
    m[2,3] <- cont_table_list$t_fanaerobic_DOWN[1,1] + 
      cont_table_list$t_unassigned_DOWN[1,2]
    
    m[3,1] <- cont_table_list$t_aerobic_UP[2,1] - m[1,1]
    m[3,2] <- cont_table_list$t_anaerobic_UP[2,1] - m[2,2]
    m[3,3] <- cont_table_list$t_fanaerobic_UP[2,1] + cont_table_list$t_unassigned_UP[2,2] - 
      cont_table_list$t_fanaerobic_DOWN[1,1] - 
      cont_table_list$t_unassigned_DOWN[1,2]
    
    rownames(m) <- c("UP","DOWN","not DA")
    colnames(m) <- c("Aerobic","Anaerobic","Other")
    cont_table_list$t_anaerobic_UP
    cont_table_list$t_aerobic_DOWN
    
    statistic <- chisq.test(m+1)
    p.value = statistic$p.value
    return(data.frame("p.value" = p.value))
  },.id = "Method")
},.id = "Threshold")

ggplot(chi3_3_df,mapping = aes(x = Threshold,y = -log10(p.value), color = Method)) + 
  geom_point() +
  geom_path(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  ggtitle(label = "Chi-squared test 3x3 table" ,subtitle = "Aerobic and Anaerobic enrichment test p.values")
```


```{r, fig.height=7, fig.width=7}
# UP in Supragingival - Aerobic and DOWN in Supragingival - Anaerobic enrichment test p.values
double_test_df <- ldply(DA_list,function(percent_list){
  ldply(.data = percent_list,.fun = function(method){
    cont_table_list <- create_contingency_tables(method)
    
    p_ae_UP <- cont_table_list$f_aerobic_UP$p.value
    p_anae_DOWN <- cont_table_list$f_anaerobic_DOWN$p.value
    
    return(data.frame("p.value_Aerobic_UP" = p_ae_UP,"p.value_Anaerobic_DOWN" = p_anae_DOWN))
  },.id = "Method")
},.id = "Threshold")

double_p_value <- melt(double_test_df)
double_p_value[,"-log10(p-value)"] <- -log10(double_p_value$value)
double_p_value$`-log10(p-value)`[double_p_value$variable == "p.value_Anaerobic_DOWN"] <- -double_p_value$`-log10(p-value)`[double_p_value$variable == "p.value_Anaerobic_DOWN"]

ggplot(double_p_value,mapping = aes(x = Threshold,y = `-log10(p-value)`, color = Method)) + 
  geom_point() + 
  facet_wrap(~ variable,nrow = 2,scales = "free_y") +
  geom_line(aes(group = Method)) + 
  scale_color_manual(values = cols) + 
  labs(y = bquote(-log[10]~(p~value))) +
  scale_y_continuous(breaks = c(seq(-20,0,by = 5),seq(1,5,by = 1)),labels = c(seq(20,0,by = -5),seq(1,5,by = 1))) +
  ggtitle(label = "Fisher exact test" ,subtitle = "UP in Supragingival - Aerobic and DOWN in Supragingival - Anaerobic enrichment test p.values") +
  theme_minimal()
```

## Ranking

```{r, fig.height=6, fig.width=7}
ord <- order(ddply(ftest_df,.variables = ~ Method, function(met) mean(-log10(met$p.value)))$V1)

fig_b_fisher <- ggplot(ftest_df,mapping = aes(x = Method,y = Threshold)) + 
  geom_line(aes(size = -log10(p.value), alpha = -log10(p.value), group = Method)) +
  scale_color_manual(values = cols) +  
  scale_x_discrete(limits = levels(ftest_df$Method)[ord]) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom") + 
  ggtitle(label = "Method's ranking by Fisher's exact test",
          subtitle = "Top Threshold% taxa chosen from raw p-value list") +
  labs(y = "Threshold (%)", alpha = bquote(-log[10]~(p~value)), size = bquote(-log[10]~(p~value)))
fig_b_fisher
```

```{r, fig.height=6, fig.width=7}
ord <- order(ddply(combined_enrichment_df,.variables = ~ Method, function(met) mean(-log10(met$p.value)))$V1)

fig_b_comb <- ggplot(combined_enrichment_df,mapping = aes(x = Method,y = Threshold)) + 
  geom_line(aes(size = -log10(p.value), alpha = -log10(p.value), group = Method)) +
  scale_color_manual(values = cols) +  
  scale_x_discrete(limits = levels(ftest_df$Method)[ord]) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom") + 
  ggtitle(label = "Method's ranking by Fisher-combined enrichment p-values",
          subtitle = "Top Threshold% taxa chosen from raw p-value list") +
  labs(y = "Threshold (%)", alpha = bquote(-log[10]~(p~value)), size = bquote(-log[10]~(p~value)))
fig_b_comb
```

```{r, fig.height=7, fig.width=5}
ord <- order(ddply(sstatistic_df,.variables = ~ Method, function(met) mean(met$s.value)/(sd(met$s.value)))$V1)

fig_b_stat <- ggplot(sstatistic_df,mapping = aes(x = Method,y = Threshold, color = Method)) + 
  geom_line(aes(size = s.value, alpha = s.value, group = Method)) +
  scale_color_manual(values = cols) +  
  scale_x_discrete(limits = levels(sstatistic_df$Method)[ord]) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom") + 
  ggtitle(label = "Method's ranking by Penalized Proportion",
          subtitle = "Top Threshold% taxa chosen from raw p-value list") +
  labs(y = "Threshold (%)", alpha = "Penalized Proportion", size = "Penalized Proportion")
fig_b_stat
```

```{r, fig.height=6, fig.width=7}
ord <- order(ddply(chi3_3_df,.variables = ~ Method, function(met) mean(-log10(met$p.value)))$V1)

fig_b_chi3_3 <- ggplot(chi3_3_df,mapping = aes(x = Method,y = Threshold)) + 
  geom_line(aes(size = -log10(p.value), alpha = -log10(p.value), group = Method)) +
  scale_color_manual(values = cols) +  
  scale_x_discrete(limits = levels(ftest_df$Method)[ord]) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom") + 
  ggtitle(label = "Method's ranking by Fisher-combined enrichment p-values",
          subtitle = "Top Threshold% taxa chosen from raw p-value list") +
  labs(y = "Threshold (%)", alpha = bquote(-log[10]~(p~value)), size = bquote(-log[10]~(p~value)))
fig_b_chi3_3
```

```{r, fig.height=6, fig.width=7}
fig_b_TP_FP <- ggplot(TP_FP_df,mapping = aes(x = Method,y = Threshold)) + 
  geom_line(aes(size = s, alpha = s, group = Method)) +
  scale_color_manual(values = cols) +  
  scale_x_discrete(limits = levels(TP_FP_df$Method)[ord_fig_a]) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom") + 
  ggtitle(label = "Putative TP - Putative FP",
          subtitle = "Top Threshold% taxa chosen from raw p-value list") +
  labs(y = "Threshold (%)", alpha = "Putative TP - Putative FP", size = "Putative TP - Putative FP")
fig_b_TP_FP

fig_b_TP_FP_normalized <- ggplot(TP_FP_df,mapping = aes(x = Method,y = Threshold)) + 
  geom_line(aes(size = s_norm, alpha = s_norm, group = Method)) +
  scale_color_manual(values = cols) +  
  scale_x_discrete(limits = levels(TP_FP_df$Method)[ord_fig_a]) +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5),
        legend.position = "bottom") + 
  ggtitle(label = "Normalized Putative TP - Putative FP",
          subtitle = "Top Threshold% taxa chosen from raw p-value list") +
  labs(y = "Threshold (%)", alpha = "Normalized Putative TP - Putative FP", size = "Normalized Putative TP - Putative FP")
fig_b_TP_FP_normalized
```

```{r, fig.width=13, fig.height=16}
fig <- plot_grid(plotlist = list(plot_grid(plotlist = list(fig_a,
                                                    fig_b_TP_FP_method),
                                    align = "h",
                                    axis = "tb",
                                    labels = "auto", 
                                    ncol = 2, 
                                    rel_widths = c(1,1)),
                          plot_grid(plotlist = list(mf_aerobic,
                                                    mf_anaerobic),
                                    nrow = 1,
                                    align = "h",
                                    axis = "tb",
                                    labels = c("c","d"))),
          nrow = 2,rel_heights = c(1.2,1))
fig

svg(filename = "../../fig6_WMS_supp.svg",width = 13,height = 16)
fig
dev.off()
```



