---
title: 'Enrichment analysis on real data'
author: "Matteo Calgaro"
date: 'Compiled: `r format(Sys.Date(), "%d %B, %Y")`'
output:
  html_document:
    code_folding: hide
    df_print: kable
    number_sections: no
    theme: united
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
---

<style>
#TOC {
  top: 1%;
  opacity: 0.5;
}
#TOC:hover {
  opacity: 1;
}
</style>

```{r setup, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(phyloseq)
source("../eval_functions.R")
```

# Data loading

Biological knowledge about microbial niches has been collected in literature for some of the datasets used in the concordance analysis:
<ul>
  <li>https://www.ncbi.nlm.nih.gov/pubmed/31076212</li>
  <li>https://github.com/waldronlab/nychanesmicrobiome/blob/master/inst/extdata/biosis.tsv</li>
</ul>

```{r}
load(file = "../data/16Sdatasets_for_replicability_filtered.RData")
load(file = "../data/WMSdatasets_for_replicability_filtered.RData")
genera_methabolism <- read.table(file = "../data/genera_methabolism.tsv",sep = "\t",header = TRUE)
```

# Supragingival vs Subgingival Plaque

It is of interest to perform an enrichment analysis on the lists of differential abundant genera to study the aerobic or anaerobic methabolism of genera colonizing supragingival and subgingival plaque.

```{r}
grp1 = "Subgingival Plaque"
grp2 = "Supragingival Plaque"
ps <- ps_list_16S$Subgingival_Supragingival

ps@sam_data$grp = factor(ps@sam_data$HMP_BODY_SUBSITE,levels = c(grp1,grp2),labels = c("grp1","grp2"))

# Genera metadata about methabolism
methabolism_df <- data.frame(GENUS = ps@tax_table@.Data[,"GENUS"],row.names = names(ps@tax_table@.Data[,"GENUS"]))
index_meth <- match(methabolism_df$GENUS,genera_methabolism$Genera)
methabolism_df$Meth <- genera_methabolism$Meth[index_meth]

if(!file.exists("../data/supragingival_subgingival_DA.RData")){
  eval_methods <- oneSimRunGSOwn(physeq = ps,epsilon = 1e14, 
                                 grid.keepX = c(seq(from = ntaxa(ps)*0.1,to = ntaxa(ps),by = 10)))
  ### Songbird data preparation
  ## OTU table
  write.table(x = "#OTU ID\t",
              file = "./songbird/supragingival_subgingival_otutable.tsv",
              quote = FALSE,
              row.names = FALSE, 
              col.names = FALSE,
              eol = "")
  write.table(x = ps@otu_table, append = TRUE,
              file = "./songbird/supragingival_subgingival_otutable.tsv",
              quote = FALSE,
              sep = "\t",
              row.names = TRUE,
              col.names = TRUE)
  ## Sample data
  write.table(x = "sampleID\t",
              file = "./songbird/supragingival_subgingival_samdata.tsv",
              quote = FALSE,
              row.names = FALSE, 
              col.names = FALSE,
              eol = "")
  write.table(x = ps@sam_data, 
              append = TRUE,
              file = "./songbird/supragingival_subgingival_samdata.tsv",
              quote = FALSE,
              sep = "\t",
              row.names = TRUE,
              col.names = TRUE)
  
  ## Then switch to unix bash to run songbird:
  # module load Anaconda/2019.07
  # source activate songbird_env
  # biom convert -i './songbird/supragingival_subgingival_otutable.tsv' -o './songbird/supragingival_subgingival_otutable.biom' --to-hdf5 
  # songbird multinomial --input-biom './songbird/supragingival_subgingival/supragingival_subgingival_otutable.biom' \
  #                      --metadata-file './songbird/supragingival_subgingival/supragingival_subgingival_samdata.tsv' \
  #                      --formula "grp" \
  #                      --summary-dir './songbird/supragingival_subgingival/'
  ## This procedure generates the "differentials.tsv" file
  
  songbird <- read.table(file = "./songbird/differentials.tsv",header = TRUE)
  eval_methods$songbird$statInfo <- data.frame(songbird,row.names = songbird$featureid)
  
  save(eval_methods, file = "../data/supragingival_subgingival_DA.RData")
} else load(file = "../data/supragingival_subgingival_DA.RData")
```

Many methods, many outputs:
<ul>
  <li>MAST, edgeR, limma and DESeq2 based methods have "logFC" column in statInfo matrix;</li>
  <li>metagenomeSeq has "coef_grp2" column of statInfo matrix which represents the FC;</li>
  <li>scde has "ce" column of statInfo matrix which represents the conservative estimate of expression-fold change (equals to the min(abs(c(lb, ub))), or 0 if the CI crosses the 0;</li>
  <li>seurat has "avg_logFC" column in statInfo matrix;</li>
  <li>ALDEx2 has "effect" column in statInfo matrix which represents a vector containing the per-feature effect size (grp1 vs grp2);</li>
  <li>corncob methods has "mu.grpgrp2" coefficient which represents the FC in statInfo matrix;</li>
  <li>mixMC has the "value.var" column which represents the loading;</li>
  <li>songbird has the "grp.T.grp2." column which contains the differentials for grp2.</li>
</ul>

For most methods a \code{pValMat} object is present. We use statInfo information to assess the differential abundance (DA) direction, while we use pValMat to count the number of DA features (corrected p-values).

```{r}
col_names <- c("logFC","coef_grp2","mu.grpgrp2","ce","avg_logFC","effect","value.var","grp.T.grp2.","coef")

extract_statistics <- function(method, col_names = col_names, alpha = 0.05, percent = 30, methabolism_df = methabolism_df){
  which_col <- which(col_names %in% colnames(method$statInfo))
  if("pValMat" %in% names(method)){# for most methods
    
    # pValMat investigation for DA identification
    DA_features_index <- which(method$pValMat[,"adjP"] < alpha)
    DA_features <- names(DA_features_index)
    if(is.null(rownames(method$statInfo)))
      rownames(method$statInfo) <- rownames(method$pValMat)
    
  } else if(col_names[which_col] == "grp.T.grp2."){# songbird
    
    differentials <- method$statInfo[,c("featureid",col_names[which_col])]
    ordered_index <- order(-abs(differentials[,col_names[which_col]]))
    DA_features_index <- ordered_index[1:round(nrow(differentials)*percent/100)] # Top 10% of total taxa
    DA_features <- differentials[DA_features_index,"featureid"]
    
  } else if(col_names[which_col] == "value.var"){# mixMC
    
    DA_features <- rownames(method$statInfo)
    
  } 
  
  if(col_names[which_col] == "coef"){# corncob methods have melted statInfo matrix
    
    which_col <- which(col_names %in% method$statInfo$coef)
    method$statInfo <- method$statInfo[method$statInfo$coef == col_names[which_col],c("feature","Estimate")]
    colnames(method$statInfo) <- c("feature",col_names[which_col])
    rownames(method$statInfo) <- method$statInfo$feature
    
  }
  
  if(col_names[which_col] == "coef_grp2"){# metagenomeSeq has the log-odd values which have the opposite sign
    
    method$statInfo[,col_names[which_col]] <- -method$statInfo[,col_names[which_col]] 
    
  }
  
  # statInfo investigation for DA direction
  direction <- sign(method$statInfo[DA_features,col_names[which_col]])
  direction <- factor(as.character(direction),levels = c("-1","1"),labels = c("DOWN","UP"))
  # association with genus methabolism
  meth = methabolism_df[DA_features,"Meth"]
  
  return(data.frame("DA_features" = DA_features,"DA_Direction" = direction,"Methabolism" = meth))
}
```

```{r}
DA_methods <- lapply(eval_methods,function(method) 
  extract_statistics(method = method,
                     col_names = col_names,
                     alpha = 0.5,percent = 30,
                     methabolism_df = methabolism_df))
```

```{r}
create_contingency_tables <- function(method){
  method$Anaerobic <- "non-Anaerobic"
  method$Anaerobic[method$Methabolism == "Anaerobic"] <- "Anaerobic"
  
  method$Aerobic <- "non-Aerobic"
  method$Aerobic[method$Methabolism == "Aerobic"] <- "Aerobic"
  
  t_anaerobic <- table(method$DA_Direction,method$Anaerobic)
  t_aerobic <- table(method$DA_Direction,method$Aerobic)
  
  if(ncol(t_anaerobic) == 1){
    t <- as.table(matrix(c(0,0,0,0),nrow = 2,ncol = 2))
    rownames(t) <- c("DOWN","UP")
    colnames(t) <- c("Anaerobic","non-Anaerobic")
    t[,colnames(t_anaerobic)] <- t_anaerobic
    t_anaerobic <- t
  } 
  if(ncol(t_aerobic) == 1){
    t <- as.table(matrix(c(0,0,0,0),nrow = 2,ncol = 2))
    rownames(t) <- c("DOWN","UP")
    colnames(t) <- c("Aerobic","non-Aerobic")
    t[,colnames(t_aerobic)] <- t_aerobic
    t_aerobic <- t 
  } 
  
  f_anaerobic <- fisher.test(t_anaerobic)
  f_aerobic <- fisher.test(t_aerobic)
  return(list("Aerobic" = t_aerobic ,"Anaerobic" = t_anaerobic ,"Fisher Aerobic" = f_aerobic,"Fisher Anaerobic" = f_anaerobic))
}
```

```{r}
contingency_table_list <- lapply(DA_methods,create_contingency_tables)

# Extract counts
contingency_table_df <- ldply(contingency_table_list,.fun = function(method) cbind(method$Aerobic,method$Anaerobic))
colnames(contingency_table_df)[1] <- "Method"
contingency_table_df$Direction <- c("DOWN","UP")
contingency_table_df$Direction <- factor(contingency_table_df$Direction)
contingency_table_melted <- melt(contingency_table_df)

Aerobic_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) 
  c(method$`Fisher Aerobic`$conf.int[1:2],method$`Fisher Aerobic`$estimate[1],"Aerobic"))
Anaerobic_odds_ratio_df <- ldply(contingency_table_list,.fun = function(method) 
  c(method$`Fisher Anaerobic`$conf.int[1:2],method$`Fisher Anaerobic`$estimate[1],"Anaerobic"))
colnames(Aerobic_odds_ratio_df) <- colnames(Anaerobic_odds_ratio_df) <- c("Method","LowerBound","UpperBound","Odds ratio","Methabolism")

odds_ratio_df <- rbind(Aerobic_odds_ratio_df,Anaerobic_odds_ratio_df)
odds_ratio_melted <- melt(odds_ratio_df,id.vars = c("Method","Methabolism"))
```

```{r}
plot_contingency_table <- function(methabolism, method_name){
  ggplot(data = contingency_table_melted[contingency_table_melted$variable %in% c(methabolism,paste0("non-",methabolism)) & contingency_table_melted$Method == method_name,],mapping = aes(x = variable, y = Direction, fill = value)) + 
    geom_tile(width = 0.8, height = 0.8) +
    geom_label(aes(label = value), fill = "white") +
    coord_fixed() + 
    theme_minimal() +
    scale_y_discrete(limits = c("UP","DOWN")) +
    scale_x_discrete(position = "top") +
    ylab(label = element_blank()) +
    theme(legend.position = "none",
          axis.title.x = element_blank()) +
    ggtitle(label = element_blank(),subtitle = method_name)
}
```

```{r, fig.height=7, fig.width=16}
plot_list_Aerobic <- lapply(names(contingency_table_list),function(method) plot_contingency_table(methabolism = "Aerobic",method_name = method))

plot_list_Anaerobic <- lapply(names(contingency_table_list),function(method) plot_contingency_table(methabolism = "Anaerobic",method_name = method))

library(cowplot)
plot_grid(plotlist = plot_list_Aerobic,nrow = 3, ncol = 6)
plot_grid(plotlist = plot_list_Anaerobic,nrow = 3, ncol = 6)
```

```{r, fig.width=10, fig.height=7}
odds_ratio_df$`Odds ratio` <- as.numeric(odds_ratio_df$`Odds ratio`)
odds_ratio_df$LowerBound <- as.numeric(odds_ratio_df$LowerBound)
odds_ratio_df$UpperBound <- as.numeric(odds_ratio_df$UpperBound)
odds_ratio_df$UpperBoundLimited <- odds_ratio_df$UpperBound
odds_ratio_df$UpperBoundLimited[odds_ratio_df$UpperBound == Inf] <- max(odds_ratio_df$UpperBound[odds_ratio_df$UpperBound!=Inf])

odds_ratio_df$Method <- factor(odds_ratio_df$Method)

library(ggforce)

ord <- order(ddply(odds_ratio_df,.variables = ~ Method, function(met) met[met$Methabolism == "Aerobic","Odds ratio"])$V1)

svg(filename = "../../fig_odds.svg",width = 10,height = 7)
ggplot(data = odds_ratio_df ,mapping = aes(x = Method, y = log(`Odds ratio`), color = Methabolism)) + 
  theme_grey() +
  theme(axis.text.x = element_text(angle = 90,hjust = 1),
        axis.text.y.right = element_text(angle = 90,hjust = 0.5),
        axis.title.y.right = element_text(angle = 90),
        axis.title.x = element_text(angle = 180)) +
  geom_point(position = position_dodge(width = 0.1)) +
  geom_errorbar(mapping = aes(ymin = log(LowerBound), ymax = log(UpperBoundLimited))) +
  scale_x_discrete(limits = levels(odds_ratio_df$Method)[ord]) +
  scale_y_continuous(position = "right") +
  ylab("Log Odds Ratio") +
  #facet_zoom(ylim = c(-4,4),split = TRUE) +
  geom_hline(mapping = aes(yintercept = 0), color = "black", lty = 2)
dev.off()
```









